{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"gpuType":"T4","authorship_tag":"ABX9TyMdCijfaQNWcq+QVHuVoPKl"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"accelerator":"GPU"},"cells":[{"cell_type":"markdown","source":["# **Step 1: Installation**"],"metadata":{"id":"xALbPRj-H5TD"}},{"cell_type":"code","source":["!python --version"],"metadata":{"id":"lD1CqyWnIAht","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1729565357918,"user_tz":480,"elapsed":499,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"b3356fdf-a0fd-4178-b7cd-94f84284bf85"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Python 3.10.12\n"]}]},{"cell_type":"code","source":["!apt install python3.10-venv"],"metadata":{"id":"use0tlg7IFIT","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1729573938457,"user_tz":480,"elapsed":8424,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"2991d732-d4ca-4190-c43b-6db4c58f0a56"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Reading package lists... Done\n","Building dependency tree... Done\n","Reading state information... Done\n","The following additional packages will be installed:\n","  python3-pip-whl python3-setuptools-whl\n","The following NEW packages will be installed:\n","  python3-pip-whl python3-setuptools-whl python3.10-venv\n","0 upgraded, 3 newly installed, 0 to remove and 49 not upgraded.\n","Need to get 2,474 kB of archives.\n","After this operation, 2,885 kB of additional disk space will be used.\n","Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3-pip-whl all 22.0.2+dfsg-1ubuntu0.4 [1,680 kB]\n","Get:2 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3-setuptools-whl all 59.6.0-1.2ubuntu0.22.04.2 [788 kB]\n","Get:3 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 python3.10-venv amd64 3.10.12-1~22.04.6 [5,722 B]\n","Fetched 2,474 kB in 3s (974 kB/s)\n","Selecting previously unselected package python3-pip-whl.\n","(Reading database ... 123629 files and directories currently installed.)\n","Preparing to unpack .../python3-pip-whl_22.0.2+dfsg-1ubuntu0.4_all.deb ...\n","Unpacking python3-pip-whl (22.0.2+dfsg-1ubuntu0.4) ...\n","Selecting previously unselected package python3-setuptools-whl.\n","Preparing to unpack .../python3-setuptools-whl_59.6.0-1.2ubuntu0.22.04.2_all.deb ...\n","Unpacking python3-setuptools-whl (59.6.0-1.2ubuntu0.22.04.2) ...\n","Selecting previously unselected package python3.10-venv.\n","Preparing to unpack .../python3.10-venv_3.10.12-1~22.04.6_amd64.deb ...\n","Unpacking python3.10-venv (3.10.12-1~22.04.6) ...\n","Setting up python3-setuptools-whl (59.6.0-1.2ubuntu0.22.04.2) ...\n","Setting up python3-pip-whl (22.0.2+dfsg-1ubuntu0.4) ...\n","Setting up python3.10-venv (3.10.12-1~22.04.6) ...\n"]}]},{"cell_type":"code","source":["!python -m venv ./venv"],"metadata":{"id":"A05o5tDtIr87"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!source ./venv/bin/activate"],"metadata":{"id":"X7zWWdUnIssI"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!pip install --upgrade pip setuptools wheel --no-cache-dir --force-reinstall"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":620},"id":"jIYZpMMNi4Ky","executionInfo":{"status":"ok","timestamp":1729573979292,"user_tz":480,"elapsed":11194,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"07406747-28c1-4b66-b875-047bdd301918"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting pip\n","  Downloading pip-24.2-py3-none-any.whl.metadata (3.6 kB)\n","Collecting setuptools\n","  Downloading setuptools-75.2.0-py3-none-any.whl.metadata (6.9 kB)\n","Collecting wheel\n","  Downloading wheel-0.44.0-py3-none-any.whl.metadata (2.3 kB)\n","Downloading pip-24.2-py3-none-any.whl (1.8 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.8/1.8 MB\u001b[0m \u001b[31m201.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading setuptools-75.2.0-py3-none-any.whl (1.2 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m358.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading wheel-0.44.0-py3-none-any.whl (67 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.1/67.1 kB\u001b[0m \u001b[31m296.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: wheel, setuptools, pip\n","  Attempting uninstall: wheel\n","    Found existing installation: wheel 0.44.0\n","    Uninstalling wheel-0.44.0:\n","      Successfully uninstalled wheel-0.44.0\n","  Attempting uninstall: setuptools\n","    Found existing installation: setuptools 75.1.0\n","    Uninstalling setuptools-75.1.0:\n","      Successfully uninstalled setuptools-75.1.0\n","  Attempting uninstall: pip\n","    Found existing installation: pip 24.1.2\n","    Uninstalling pip-24.1.2:\n","      Successfully uninstalled pip-24.1.2\n","\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n","ipython 7.34.0 requires jedi>=0.16, which is not installed.\u001b[0m\u001b[31m\n","\u001b[0mSuccessfully installed pip-24.2 setuptools-75.2.0 wheel-0.44.0\n"]},{"output_type":"display_data","data":{"application/vnd.colab-display-data+json":{"pip_warning":{"packages":["_distutils_hack","pkg_resources","setuptools"]},"id":"291dcf57136d4866a7c1823201cc7d3e"}},"metadata":{}}]},{"cell_type":"code","source":["!pip install --upgrade pip setuptools wheel --use-deprecated=legacy-resolver"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-fdW9jQ1I1vT","executionInfo":{"status":"ok","timestamp":1729572976134,"user_tz":480,"elapsed":4712,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"fb7ef003-95fb-4ec3-aa11-f869b7bf2ac6"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already up-to-date: pip in /usr/local/lib/python3.10/dist-packages (24.2)\n","Requirement already up-to-date: setuptools in /usr/local/lib/python3.10/dist-packages (75.2.0)\n","Requirement already up-to-date: wheel in /usr/local/lib/python3.10/dist-packages (0.44.0)\n"]}]},{"cell_type":"code","source":["!pip install -U cellxgene-census --no-cache-dir --force-reinstall --use-deprecated=legacy-resolver"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"id":"cZ0NVRduJiyr","executionInfo":{"status":"ok","timestamp":1729574146625,"user_tz":480,"elapsed":65716,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"fff92d2c-a0b0-44f2-f605-76e965d6ff21"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting cellxgene-census\n","  Downloading cellxgene_census-1.16.2-py3-none-any.whl (54 kB)\n","Collecting tiledbsoma!=1.14.1,>=1.12.3 (from cellxgene-census)\n","  Downloading tiledbsoma-1.14.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.5 MB)\n","\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/17.5 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.5/17.5 MB\u001b[0m \u001b[31m232.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting anndata (from cellxgene-census)\n","  Downloading anndata-0.10.9-py3-none-any.whl (128 kB)\n","Collecting numpy<2.0,>=1.23 (from cellxgene-census)\n","  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.2/18.2 MB\u001b[0m \u001b[31m547.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting requests (from cellxgene-census)\n","  Downloading requests-2.32.3-py3-none-any.whl (64 kB)\n","Collecting typing-extensions (from cellxgene-census)\n","  Downloading typing_extensions-4.12.2-py3-none-any.whl (37 kB)\n","Collecting s3fs>=2021.06.1 (from cellxgene-census)\n","  Downloading s3fs-2024.10.0-py3-none-any.whl (29 kB)\n","Collecting attrs>=22.2 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading attrs-24.2.0-py3-none-any.whl (63 kB)\n","Collecting numba>=0.58.0 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading numba-0.60.0-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (3.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.7/3.7 MB\u001b[0m \u001b[31m188.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pandas (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.1/13.1 MB\u001b[0m \u001b[31m524.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pyarrow (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pyarrow-17.0.0-cp310-cp310-manylinux_2_28_x86_64.whl (39.9 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m39.9/39.9 MB\u001b[0m \u001b[31m534.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scanpy>=1.9.2 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading scanpy-1.10.3-py3-none-any.whl (2.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m561.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scipy (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading scipy-1.14.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (41.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.2/41.2 MB\u001b[0m \u001b[31m399.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting somacore==1.0.17 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading somacore-1.0.17-py3-none-any.whl (38 kB)\n","Collecting tiledb~=0.32.0 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading tiledb-0.32.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m16.8/16.8 MB\u001b[0m \u001b[31m388.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting array-api-compat!=1.5,>1.4 (from anndata->cellxgene-census)\n","  Downloading array_api_compat-1.9-py3-none-any.whl (49 kB)\n","Collecting exceptiongroup; python_version < \"3.11\" (from anndata->cellxgene-census)\n","  Downloading exceptiongroup-1.2.2-py3-none-any.whl (16 kB)\n","Collecting h5py>=3.1 (from anndata->cellxgene-census)\n","  Downloading h5py-3.12.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.3/5.3 MB\u001b[0m \u001b[31m298.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting natsort (from anndata->cellxgene-census)\n","  Downloading natsort-8.4.0-py3-none-any.whl (38 kB)\n","Collecting packaging>=20.0 (from anndata->cellxgene-census)\n","  Downloading packaging-24.1-py3-none-any.whl (53 kB)\n","Collecting charset-normalizer<4,>=2 (from requests->cellxgene-census)\n","  Downloading charset_normalizer-3.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (144 kB)\n","Collecting idna<4,>=2.5 (from requests->cellxgene-census)\n","  Downloading idna-3.10-py3-none-any.whl (70 kB)\n","Collecting urllib3<3,>=1.21.1 (from requests->cellxgene-census)\n","  Downloading urllib3-2.2.3-py3-none-any.whl (126 kB)\n","Collecting certifi>=2017.4.17 (from requests->cellxgene-census)\n","  Downloading certifi-2024.8.30-py3-none-any.whl (167 kB)\n","Collecting aiobotocore<3.0.0,>=2.5.4 (from s3fs>=2021.06.1->cellxgene-census)\n","  Downloading aiobotocore-2.15.2-py3-none-any.whl (77 kB)\n","Collecting fsspec==2024.10.0.* (from s3fs>=2021.06.1->cellxgene-census)\n","  Downloading fsspec-2024.10.0-py3-none-any.whl (179 kB)\n","Collecting aiohttp!=4.0.0a0,!=4.0.0a1 (from s3fs>=2021.06.1->cellxgene-census)\n","  Downloading aiohttp-3.10.10-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m411.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting llvmlite<0.44,>=0.43.0dev0 (from numba>=0.58.0->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading llvmlite-0.43.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (43.9 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m43.9/43.9 MB\u001b[0m \u001b[31m114.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting python-dateutil>=2.8.2 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n","Collecting pytz>=2020.1 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pytz-2024.2-py2.py3-none-any.whl (508 kB)\n","Collecting tzdata>=2022.7 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading tzdata-2024.2-py2.py3-none-any.whl (346 kB)\n","Collecting joblib (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading joblib-1.4.2-py3-none-any.whl (301 kB)\n","Collecting legacy-api-wrap>=1.4 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading legacy_api_wrap-1.4-py3-none-any.whl (15 kB)\n","Collecting matplotlib>=3.6 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading matplotlib-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.3/8.3 MB\u001b[0m \u001b[31m330.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting networkx>=2.7 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading networkx-3.4.2-py3-none-any.whl (1.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m422.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting patsy (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading patsy-0.5.6-py2.py3-none-any.whl (233 kB)\n","Collecting pynndescent>=0.5 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pynndescent-0.5.13-py3-none-any.whl (56 kB)\n","Collecting scikit-learn>=0.24 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading scikit_learn-1.5.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.3/13.3 MB\u001b[0m \u001b[31m365.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting seaborn>=0.13 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading seaborn-0.13.2-py3-none-any.whl (294 kB)\n","Collecting session-info (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading session_info-1.0.0.tar.gz (24 kB)\n","  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","Collecting statsmodels>=0.13 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading statsmodels-0.14.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.8/10.8 MB\u001b[0m \u001b[31m373.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting tqdm (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading tqdm-4.66.5-py3-none-any.whl (78 kB)\n","Collecting umap-learn!=0.5.0,>=0.5 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading umap_learn-0.5.6-py3-none-any.whl (85 kB)\n","Collecting pyarrow-hotfix (from somacore==1.0.17->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pyarrow_hotfix-0.6-py3-none-any.whl (7.9 kB)\n","Collecting botocore<1.35.37,>=1.35.16 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading botocore-1.35.36-py3-none-any.whl (12.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.6/12.6 MB\u001b[0m \u001b[31m472.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting wrapt<2.0.0,>=1.10.10 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading wrapt-1.16.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (80 kB)\n","Collecting aioitertools<1.0.0,>=0.5.1 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading aioitertools-0.12.0-py3-none-any.whl (24 kB)\n","Collecting aiohappyeyeballs>=2.3.0 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading aiohappyeyeballs-2.4.3-py3-none-any.whl (14 kB)\n","Collecting aiosignal>=1.1.2 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading aiosignal-1.3.1-py3-none-any.whl (7.6 kB)\n","Collecting frozenlist>=1.1.1 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading frozenlist-1.4.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (239 kB)\n","Collecting multidict<7.0,>=4.5 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading multidict-6.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (124 kB)\n","Collecting yarl<2.0,>=1.12.0 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading yarl-1.16.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (318 kB)\n","Collecting async-timeout<5.0,>=4.0; python_version < \"3.11\" (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading async_timeout-4.0.3-py3-none-any.whl (5.7 kB)\n","Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)\n","Collecting contourpy>=1.0.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading contourpy-1.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (322 kB)\n","Collecting cycler>=0.10 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\n","Collecting fonttools>=4.22.0 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading fonttools-4.54.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.6/4.6 MB\u001b[0m \u001b[31m428.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting kiwisolver>=1.3.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading kiwisolver-1.4.7-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m404.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pillow>=8 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pillow-11.0.0-cp310-cp310-manylinux_2_28_x86_64.whl (4.4 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.4/4.4 MB\u001b[0m \u001b[31m357.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pyparsing>=2.3.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading pyparsing-3.2.0-py3-none-any.whl (106 kB)\n","Collecting threadpoolctl>=3.1.0 (from scikit-learn>=0.24->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading threadpoolctl-3.5.0-py3-none-any.whl (18 kB)\n","Collecting stdlib_list (from session-info->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census)\n","  Downloading stdlib_list-0.11.0-py3-none-any.whl (83 kB)\n","Collecting jmespath<2.0.0,>=0.7.1 (from botocore<1.35.37,>=1.35.16->aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading jmespath-1.0.1-py3-none-any.whl (20 kB)\n","Collecting propcache>=0.2.0 (from yarl<2.0,>=1.12.0->aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census)\n","  Downloading propcache-0.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (208 kB)\n","Building wheels for collected packages: session-info\n","  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8024 sha256=7c65f43296ddede928bb70db206ce9291fae9036e4a6903e91273ed02e59ef41\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-s1mghubu/wheels/6a/aa/b9/eb5d4031476ec10802795b97ccf937b9bd998d68a9b268765a\n","Successfully built session-info\n","Installing collected packages: array-api-compat, exceptiongroup, numpy, h5py, natsort, packaging, six, python-dateutil, pytz, tzdata, pandas, scipy, anndata, attrs, llvmlite, numba, pyarrow, joblib, legacy-api-wrap, contourpy, cycler, fonttools, kiwisolver, pillow, pyparsing, matplotlib, networkx, patsy, threadpoolctl, scikit-learn, pynndescent, seaborn, stdlib_list, session-info, statsmodels, tqdm, umap-learn, scanpy, pyarrow-hotfix, typing-extensions, somacore, tiledb, tiledbsoma, charset-normalizer, idna, urllib3, certifi, requests, jmespath, botocore, aiohappyeyeballs, frozenlist, aiosignal, multidict, propcache, yarl, async-timeout, aiohttp, wrapt, aioitertools, aiobotocore, fsspec, s3fs, cellxgene-census\n","  Attempting uninstall: array-api-compat\n","    Found existing installation: array_api_compat 1.9\n","    Uninstalling array_api_compat-1.9:\n","      Successfully uninstalled array_api_compat-1.9\n","  Attempting uninstall: exceptiongroup\n","    Found existing installation: exceptiongroup 1.2.2\n","    Uninstalling exceptiongroup-1.2.2:\n","      Successfully uninstalled exceptiongroup-1.2.2\n","  Attempting uninstall: numpy\n","    Found existing installation: numpy 1.26.4\n","    Uninstalling numpy-1.26.4:\n","      Successfully uninstalled numpy-1.26.4\n","  Attempting uninstall: h5py\n","    Found existing installation: h5py 3.11.0\n","    Uninstalling h5py-3.11.0:\n","      Successfully uninstalled h5py-3.11.0\n","  Attempting uninstall: natsort\n","    Found existing installation: natsort 8.4.0\n","    Uninstalling natsort-8.4.0:\n","      Successfully uninstalled natsort-8.4.0\n","  Attempting uninstall: packaging\n","    Found existing installation: packaging 24.1\n","    Uninstalling packaging-24.1:\n","      Successfully uninstalled packaging-24.1\n","  Attempting uninstall: six\n","    Found existing installation: six 1.16.0\n","    Uninstalling six-1.16.0:\n","      Successfully uninstalled six-1.16.0\n","  Attempting uninstall: python-dateutil\n","    Found existing installation: python-dateutil 2.8.2\n","    Uninstalling python-dateutil-2.8.2:\n","      Successfully uninstalled python-dateutil-2.8.2\n","  Attempting uninstall: pytz\n","    Found existing installation: pytz 2024.2\n","    Uninstalling pytz-2024.2:\n","      Successfully uninstalled pytz-2024.2\n","  Attempting uninstall: tzdata\n","    Found existing installation: tzdata 2024.2\n","    Uninstalling tzdata-2024.2:\n","      Successfully uninstalled tzdata-2024.2\n","  Attempting uninstall: pandas\n","    Found existing installation: pandas 2.2.2\n","    Uninstalling pandas-2.2.2:\n","      Successfully uninstalled pandas-2.2.2\n","  Attempting uninstall: scipy\n","    Found existing installation: scipy 1.13.1\n","    Uninstalling scipy-1.13.1:\n","      Successfully uninstalled scipy-1.13.1\n","  Attempting uninstall: anndata\n","    Found existing installation: anndata 0.10.9\n","    Uninstalling anndata-0.10.9:\n","      Successfully uninstalled anndata-0.10.9\n","  Attempting uninstall: attrs\n","    Found existing installation: attrs 24.2.0\n","    Uninstalling attrs-24.2.0:\n","      Successfully uninstalled attrs-24.2.0\n","  Attempting uninstall: llvmlite\n","    Found existing installation: llvmlite 0.43.0\n","    Uninstalling llvmlite-0.43.0:\n","      Successfully uninstalled llvmlite-0.43.0\n","  Attempting uninstall: numba\n","    Found existing installation: numba 0.60.0\n","    Uninstalling numba-0.60.0:\n","      Successfully uninstalled numba-0.60.0\n","  Attempting uninstall: pyarrow\n","    Found existing installation: pyarrow 16.1.0\n","    Uninstalling pyarrow-16.1.0:\n","      Successfully uninstalled pyarrow-16.1.0\n","  Attempting uninstall: joblib\n","    Found existing installation: joblib 1.4.2\n","    Uninstalling joblib-1.4.2:\n","      Successfully uninstalled joblib-1.4.2\n","  Attempting uninstall: legacy-api-wrap\n","    Found existing installation: legacy-api-wrap 1.4\n","    Uninstalling legacy-api-wrap-1.4:\n","      Successfully uninstalled legacy-api-wrap-1.4\n","  Attempting uninstall: contourpy\n","    Found existing installation: contourpy 1.3.0\n","    Uninstalling contourpy-1.3.0:\n","      Successfully uninstalled contourpy-1.3.0\n","  Attempting uninstall: cycler\n","    Found existing installation: cycler 0.12.1\n","    Uninstalling cycler-0.12.1:\n","      Successfully uninstalled cycler-0.12.1\n","  Attempting uninstall: fonttools\n","    Found existing installation: fonttools 4.54.1\n","    Uninstalling fonttools-4.54.1:\n","      Successfully uninstalled fonttools-4.54.1\n","  Attempting uninstall: kiwisolver\n","    Found existing installation: kiwisolver 1.4.7\n","    Uninstalling kiwisolver-1.4.7:\n","      Successfully uninstalled kiwisolver-1.4.7\n","  Attempting uninstall: pillow\n","    Found existing installation: pillow 10.4.0\n","    Uninstalling pillow-10.4.0:\n","      Successfully uninstalled pillow-10.4.0\n","  Attempting uninstall: pyparsing\n","    Found existing installation: pyparsing 3.2.0\n","    Uninstalling pyparsing-3.2.0:\n","      Successfully uninstalled pyparsing-3.2.0\n","  Attempting uninstall: matplotlib\n","    Found existing installation: matplotlib 3.7.1\n","    Uninstalling matplotlib-3.7.1:\n","      Successfully uninstalled matplotlib-3.7.1\n","  Attempting uninstall: networkx\n","    Found existing installation: networkx 3.4.1\n","    Uninstalling networkx-3.4.1:\n","      Successfully uninstalled networkx-3.4.1\n","  Attempting uninstall: patsy\n","    Found existing installation: patsy 0.5.6\n","    Uninstalling patsy-0.5.6:\n","      Successfully uninstalled patsy-0.5.6\n","  Attempting uninstall: threadpoolctl\n","    Found existing installation: threadpoolctl 3.5.0\n","    Uninstalling threadpoolctl-3.5.0:\n","      Successfully uninstalled threadpoolctl-3.5.0\n","  Attempting uninstall: scikit-learn\n","    Found existing installation: scikit-learn 1.5.2\n","    Uninstalling scikit-learn-1.5.2:\n","      Successfully uninstalled scikit-learn-1.5.2\n","  Attempting uninstall: pynndescent\n","    Found existing installation: pynndescent 0.5.13\n","    Uninstalling pynndescent-0.5.13:\n","      Successfully uninstalled pynndescent-0.5.13\n","  Attempting uninstall: seaborn\n","    Found existing installation: seaborn 0.13.2\n","    Uninstalling seaborn-0.13.2:\n","      Successfully uninstalled seaborn-0.13.2\n","  Attempting uninstall: stdlib_list\n","    Found existing installation: stdlib-list 0.11.0\n","    Uninstalling stdlib-list-0.11.0:\n","      Successfully uninstalled stdlib-list-0.11.0\n","  Attempting uninstall: session-info\n","    Found existing installation: session_info 1.0.0\n","    Uninstalling session_info-1.0.0:\n","      Successfully uninstalled session_info-1.0.0\n","  Attempting uninstall: statsmodels\n","    Found existing installation: statsmodels 0.14.4\n","    Uninstalling statsmodels-0.14.4:\n","      Successfully uninstalled statsmodels-0.14.4\n","  Attempting uninstall: tqdm\n","    Found existing installation: tqdm 4.66.5\n","    Uninstalling tqdm-4.66.5:\n","      Successfully uninstalled tqdm-4.66.5\n","  Attempting uninstall: umap-learn\n","    Found existing installation: umap-learn 0.5.6\n","    Uninstalling umap-learn-0.5.6:\n","      Successfully uninstalled umap-learn-0.5.6\n","  Attempting uninstall: scanpy\n","    Found existing installation: scanpy 1.10.3\n","    Uninstalling scanpy-1.10.3:\n","      Successfully uninstalled scanpy-1.10.3\n","  Attempting uninstall: pyarrow-hotfix\n","    Found existing installation: pyarrow-hotfix 0.6\n","    Uninstalling pyarrow-hotfix-0.6:\n","      Successfully uninstalled pyarrow-hotfix-0.6\n","  Attempting uninstall: typing-extensions\n","    Found existing installation: typing_extensions 4.12.2\n","    Uninstalling typing_extensions-4.12.2:\n","      Successfully uninstalled typing_extensions-4.12.2\n","  Attempting uninstall: somacore\n","    Found existing installation: somacore 1.0.17\n","    Uninstalling somacore-1.0.17:\n","      Successfully uninstalled somacore-1.0.17\n","  Attempting uninstall: tiledb\n","    Found existing installation: tiledb 0.32.3\n","    Uninstalling tiledb-0.32.3:\n","      Successfully uninstalled tiledb-0.32.3\n","  Attempting uninstall: tiledbsoma\n","    Found existing installation: tiledbsoma 1.14.3\n","    Uninstalling tiledbsoma-1.14.3:\n","      Successfully uninstalled tiledbsoma-1.14.3\n","  Attempting uninstall: charset-normalizer\n","    Found existing installation: charset-normalizer 3.4.0\n","    Uninstalling charset-normalizer-3.4.0:\n","      Successfully uninstalled charset-normalizer-3.4.0\n","  Attempting uninstall: idna\n","    Found existing installation: idna 3.10\n","    Uninstalling idna-3.10:\n","      Successfully uninstalled idna-3.10\n","  Attempting uninstall: urllib3\n","    Found existing installation: urllib3 2.2.3\n","    Uninstalling urllib3-2.2.3:\n","      Successfully uninstalled urllib3-2.2.3\n","  Attempting uninstall: certifi\n","    Found existing installation: certifi 2024.8.30\n","    Uninstalling certifi-2024.8.30:\n","      Successfully uninstalled certifi-2024.8.30\n","  Attempting uninstall: requests\n","    Found existing installation: requests 2.32.3\n","    Uninstalling requests-2.32.3:\n","      Successfully uninstalled requests-2.32.3\n","  Attempting uninstall: jmespath\n","    Found existing installation: jmespath 1.0.1\n","    Uninstalling jmespath-1.0.1:\n","      Successfully uninstalled jmespath-1.0.1\n","  Attempting uninstall: botocore\n","    Found existing installation: botocore 1.35.36\n","    Uninstalling botocore-1.35.36:\n","      Successfully uninstalled botocore-1.35.36\n","  Attempting uninstall: aiohappyeyeballs\n","    Found existing installation: aiohappyeyeballs 2.4.3\n","    Uninstalling aiohappyeyeballs-2.4.3:\n","      Successfully uninstalled aiohappyeyeballs-2.4.3\n","  Attempting uninstall: frozenlist\n","    Found existing installation: frozenlist 1.4.1\n","    Uninstalling frozenlist-1.4.1:\n","      Successfully uninstalled frozenlist-1.4.1\n","  Attempting uninstall: aiosignal\n","    Found existing installation: aiosignal 1.3.1\n","    Uninstalling aiosignal-1.3.1:\n","      Successfully uninstalled aiosignal-1.3.1\n","  Attempting uninstall: multidict\n","    Found existing installation: multidict 6.1.0\n","    Uninstalling multidict-6.1.0:\n","      Successfully uninstalled multidict-6.1.0\n","  Attempting uninstall: propcache\n","    Found existing installation: propcache 0.2.0\n","    Uninstalling propcache-0.2.0:\n","      Successfully uninstalled propcache-0.2.0\n","  Attempting uninstall: yarl\n","    Found existing installation: yarl 1.15.2\n","    Uninstalling yarl-1.15.2:\n","      Successfully uninstalled yarl-1.15.2\n","  Attempting uninstall: async-timeout\n","    Found existing installation: async-timeout 4.0.3\n","    Uninstalling async-timeout-4.0.3:\n","      Successfully uninstalled async-timeout-4.0.3\n","  Attempting uninstall: aiohttp\n","    Found existing installation: aiohttp 3.10.10\n","    Uninstalling aiohttp-3.10.10:\n","      Successfully uninstalled aiohttp-3.10.10\n","  Attempting uninstall: wrapt\n","    Found existing installation: wrapt 1.16.0\n","    Uninstalling wrapt-1.16.0:\n","      Successfully uninstalled wrapt-1.16.0\n","  Attempting uninstall: aioitertools\n","    Found existing installation: aioitertools 0.12.0\n","    Uninstalling aioitertools-0.12.0:\n","      Successfully uninstalled aioitertools-0.12.0\n","  Attempting uninstall: aiobotocore\n","    Found existing installation: aiobotocore 2.15.2\n","    Uninstalling aiobotocore-2.15.2:\n","      Successfully uninstalled aiobotocore-2.15.2\n","  Attempting uninstall: fsspec\n","    Found existing installation: fsspec 2024.10.0\n","    Uninstalling fsspec-2024.10.0:\n","      Successfully uninstalled fsspec-2024.10.0\n","  Attempting uninstall: s3fs\n","    Found existing installation: s3fs 2024.10.0\n","    Uninstalling s3fs-2024.10.0:\n","      Successfully uninstalled s3fs-2024.10.0\n","  Attempting uninstall: cellxgene-census\n","    Found existing installation: cellxgene-census 1.16.2\n","    Uninstalling cellxgene-census-1.16.2:\n","      Successfully uninstalled cellxgene-census-1.16.2\n","\u001b[31mERROR: pip's legacy dependency resolver does not consider dependency conflicts when selecting packages. This behaviour is the source of the following dependency conflicts.\n","cudf-cu12 24.10.1 requires pandas<2.2.3dev0,>=2.0, but you'll have pandas 2.2.3 which is incompatible.\n","dask 2024.8.2 requires cloudpickle>=3.0.0, but you'll have cloudpickle 2.2.1 which is incompatible.\n","datasets 2.21.0 requires fsspec[http]<=2024.6.1,>=2023.1.0, but you'll have fsspec 2024.10.0 which is incompatible.\n","gcsfs 2024.6.1 requires fsspec==2024.6.1, but you'll have fsspec 2024.10.0 which is incompatible.\n","gensim 4.3.3 requires scipy<1.14.0,>=1.7.0, but you'll have scipy 1.14.1 which is incompatible.\n","google-colab 1.0.0 requires pandas==2.2.2, but you'll have pandas 2.2.3 which is incompatible.\u001b[0m\u001b[31m\n","\u001b[0mSuccessfully installed aiobotocore-2.15.2 aiohappyeyeballs-2.4.3 aiohttp-3.10.10 aioitertools-0.12.0 aiosignal-1.3.1 anndata-0.10.9 array-api-compat-1.9 async-timeout-4.0.3 attrs-24.2.0 botocore-1.35.36 cellxgene-census-1.16.2 certifi-2024.8.30 charset-normalizer-3.4.0 contourpy-1.3.0 cycler-0.12.1 exceptiongroup-1.2.2 fonttools-4.54.1 frozenlist-1.4.1 fsspec-2024.10.0 h5py-3.12.1 idna-3.10 jmespath-1.0.1 joblib-1.4.2 kiwisolver-1.4.7 legacy-api-wrap-1.4 llvmlite-0.43.0 matplotlib-3.9.2 multidict-6.1.0 natsort-8.4.0 networkx-3.4.2 numba-0.60.0 numpy-1.26.4 packaging-24.1 pandas-2.2.3 patsy-0.5.6 pillow-11.0.0 propcache-0.2.0 pyarrow-17.0.0 pyarrow-hotfix-0.6 pynndescent-0.5.13 pyparsing-3.2.0 python-dateutil-2.9.0.post0 pytz-2024.2 requests-2.32.3 s3fs-2024.10.0 scanpy-1.10.3 scikit-learn-1.5.2 scipy-1.14.1 seaborn-0.13.2 session-info-1.0.0 six-1.16.0 somacore-1.0.17 statsmodels-0.14.4 stdlib_list-0.11.0 threadpoolctl-3.5.0 tiledb-0.32.3 tiledbsoma-1.14.3 tqdm-4.66.5 typing-extensions-4.12.2 tzdata-2024.2 umap-learn-0.5.6 urllib3-2.2.3 wrapt-1.16.0 yarl-1.16.0\n"]},{"output_type":"display_data","data":{"application/vnd.colab-display-data+json":{"pip_warning":{"packages":["PIL","certifi","cycler","dateutil","kiwisolver","six"]},"id":"002ee643a37a42faa5878824ada5691c"}},"metadata":{}}]},{"cell_type":"code","source":["!pip install -U cellxgene-census[experimental] --no-cache-dir --force-reinstall --use-deprecated=legacy-resolver"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"id":"2o_rLIMXKPUw","outputId":"f1d755f0-7f1b-460d-cb6f-7a3dbd6054a1","executionInfo":{"status":"ok","timestamp":1729574432792,"user_tz":480,"elapsed":261761,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting cellxgene-census[experimental]\n","  Downloading cellxgene_census-1.16.2-py3-none-any.whl (54 kB)\n","Collecting tiledbsoma!=1.14.1,>=1.12.3 (from cellxgene-census[experimental])\n","  Downloading tiledbsoma-1.14.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.5 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.5/17.5 MB\u001b[0m \u001b[31m129.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting anndata (from cellxgene-census[experimental])\n","  Downloading anndata-0.10.9-py3-none-any.whl (128 kB)\n","Collecting numpy<2.0,>=1.23 (from cellxgene-census[experimental])\n","  Downloading numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.2/18.2 MB\u001b[0m \u001b[31m481.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting requests (from cellxgene-census[experimental])\n","  Downloading requests-2.32.3-py3-none-any.whl (64 kB)\n","Collecting typing-extensions (from cellxgene-census[experimental])\n","  Downloading typing_extensions-4.12.2-py3-none-any.whl (37 kB)\n","Collecting s3fs>=2021.06.1 (from cellxgene-census[experimental])\n","  Downloading s3fs-2024.10.0-py3-none-any.whl (29 kB)\n","Collecting torch; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading torch-2.5.0-cp310-cp310-manylinux1_x86_64.whl (906.4 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m906.4/906.4 MB\u001b[0m \u001b[31m91.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting torchdata~=0.7; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading torchdata-0.9.0-cp310-cp310-manylinux1_x86_64.whl (2.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.7/2.7 MB\u001b[0m \u001b[31m612.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scikit-learn>=1.2; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading scikit_learn-1.5.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.3/13.3 MB\u001b[0m \u001b[31m244.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scikit-misc<0.4,>=0.2; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading scikit_misc-0.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.8/10.8 MB\u001b[0m \u001b[31m96.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting datasets~=2.0; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading datasets-2.21.0-py3-none-any.whl (527 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m527.3/527.3 kB\u001b[0m \u001b[31m491.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting tdigest~=0.5; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading tdigest-0.5.2.2-py2.py3-none-any.whl (9.4 kB)\n","Collecting tiledb-vector-search~=0.2; extra == \"experimental\" (from cellxgene-census[experimental])\n","  Downloading tiledb_vector_search-0.10.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (73.4 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.4/73.4 MB\u001b[0m \u001b[31m130.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting attrs>=22.2 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading attrs-24.2.0-py3-none-any.whl (63 kB)\n","Collecting numba>=0.58.0 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading numba-0.60.0-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (3.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.7/3.7 MB\u001b[0m \u001b[31m345.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pandas (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.1/13.1 MB\u001b[0m \u001b[31m373.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pyarrow (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pyarrow-17.0.0-cp310-cp310-manylinux_2_28_x86_64.whl (39.9 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m39.9/39.9 MB\u001b[0m \u001b[31m122.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scanpy>=1.9.2 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading scanpy-1.10.3-py3-none-any.whl (2.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m227.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting scipy (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading scipy-1.14.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (41.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.2/41.2 MB\u001b[0m \u001b[31m203.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting somacore==1.0.17 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading somacore-1.0.17-py3-none-any.whl (38 kB)\n","Collecting tiledb~=0.32.0 (from tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading tiledb-0.32.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m16.8/16.8 MB\u001b[0m \u001b[31m197.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting array-api-compat!=1.5,>1.4 (from anndata->cellxgene-census[experimental])\n","  Downloading array_api_compat-1.9-py3-none-any.whl (49 kB)\n","Collecting exceptiongroup; python_version < \"3.11\" (from anndata->cellxgene-census[experimental])\n","  Downloading exceptiongroup-1.2.2-py3-none-any.whl (16 kB)\n","Collecting h5py>=3.1 (from anndata->cellxgene-census[experimental])\n","  Downloading h5py-3.12.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.3/5.3 MB\u001b[0m \u001b[31m461.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting natsort (from anndata->cellxgene-census[experimental])\n","  Downloading natsort-8.4.0-py3-none-any.whl (38 kB)\n","Collecting packaging>=20.0 (from anndata->cellxgene-census[experimental])\n","  Downloading packaging-24.1-py3-none-any.whl (53 kB)\n","Collecting charset-normalizer<4,>=2 (from requests->cellxgene-census[experimental])\n","  Downloading charset_normalizer-3.4.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (144 kB)\n","Collecting idna<4,>=2.5 (from requests->cellxgene-census[experimental])\n","  Downloading idna-3.10-py3-none-any.whl (70 kB)\n","Collecting urllib3<3,>=1.21.1 (from requests->cellxgene-census[experimental])\n","  Downloading urllib3-2.2.3-py3-none-any.whl (126 kB)\n","Collecting certifi>=2017.4.17 (from requests->cellxgene-census[experimental])\n","  Downloading certifi-2024.8.30-py3-none-any.whl (167 kB)\n","Collecting aiobotocore<3.0.0,>=2.5.4 (from s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading aiobotocore-2.15.2-py3-none-any.whl (77 kB)\n","Collecting fsspec==2024.10.0.* (from s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading fsspec-2024.10.0-py3-none-any.whl (179 kB)\n","Collecting aiohttp!=4.0.0a0,!=4.0.0a1 (from s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading aiohttp-3.10.10-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m386.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting filelock (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading filelock-3.16.1-py3-none-any.whl (16 kB)\n","Collecting networkx (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading networkx-3.4.2-py3-none-any.whl (1.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m390.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting jinja2 (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading jinja2-3.1.4-py3-none-any.whl (133 kB)\n","Collecting nvidia-cuda-nvrtc-cu12==12.4.127; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cuda_nvrtc_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (24.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m24.6/24.6 MB\u001b[0m \u001b[31m437.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cuda-runtime-cu12==12.4.127; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cuda_runtime_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (883 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m883.7/883.7 kB\u001b[0m \u001b[31m469.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cuda-cupti-cu12==12.4.127; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cuda_cupti_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (13.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.8/13.8 MB\u001b[0m \u001b[31m403.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cudnn-cu12==9.1.0.70; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl (664.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m664.8/664.8 MB\u001b[0m \u001b[31m88.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cublas-cu12==12.4.5.8; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cublas_cu12-12.4.5.8-py3-none-manylinux2014_x86_64.whl (363.4 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m363.4/363.4 MB\u001b[0m \u001b[31m260.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cufft-cu12==11.2.1.3; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cufft_cu12-11.2.1.3-py3-none-manylinux2014_x86_64.whl (211.5 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m211.5/211.5 MB\u001b[0m \u001b[31m235.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-curand-cu12==10.3.5.147; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_curand_cu12-10.3.5.147-py3-none-manylinux2014_x86_64.whl (56.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.3/56.3 MB\u001b[0m \u001b[31m397.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cusolver-cu12==11.6.1.9; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cusolver_cu12-11.6.1.9-py3-none-manylinux2014_x86_64.whl (127.9 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m127.9/127.9 MB\u001b[0m \u001b[31m246.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cusparse-cu12==12.3.1.170; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_cusparse_cu12-12.3.1.170-py3-none-manylinux2014_x86_64.whl (207.5 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.5/207.5 MB\u001b[0m \u001b[31m366.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-nccl-cu12==2.21.5; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_nccl_cu12-2.21.5-py3-none-manylinux2014_x86_64.whl (188.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m188.7/188.7 MB\u001b[0m \u001b[31m365.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-nvtx-cu12==12.4.127; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_nvtx_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (99 kB)\n","Collecting nvidia-nvjitlink-cu12==12.4.127; platform_system == \"Linux\" and platform_machine == \"x86_64\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading nvidia_nvjitlink_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (21.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.1/21.1 MB\u001b[0m \u001b[31m272.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting triton==3.1.0; platform_system == \"Linux\" and platform_machine == \"x86_64\" and python_version < \"3.13\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading triton-3.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (209.5 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m209.5/209.5 MB\u001b[0m \u001b[31m230.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting sympy==1.13.1; python_version >= \"3.9\" (from torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading sympy-1.13.1-py3-none-any.whl (6.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.2/6.2 MB\u001b[0m \u001b[31m497.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting joblib>=1.2.0 (from scikit-learn>=1.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading joblib-1.4.2-py3-none-any.whl (301 kB)\n","Collecting threadpoolctl>=3.1.0 (from scikit-learn>=1.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading threadpoolctl-3.5.0-py3-none-any.whl (18 kB)\n","Collecting dill<0.3.9,>=0.3.0 (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading dill-0.3.8-py3-none-any.whl (116 kB)\n","Collecting tqdm>=4.66.3 (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading tqdm-4.66.5-py3-none-any.whl (78 kB)\n","Collecting xxhash (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading xxhash-3.5.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (194 kB)\n","Collecting multiprocess (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading multiprocess-0.70.17-py310-none-any.whl (134 kB)\n","Collecting huggingface-hub>=0.21.2 (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading huggingface_hub-0.26.1-py3-none-any.whl (447 kB)\n","Collecting pyyaml>=5.1 (from datasets~=2.0; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading PyYAML-6.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (751 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m751.2/751.2 kB\u001b[0m \u001b[31m682.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting accumulation-tree (from tdigest~=0.5; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading accumulation_tree-0.6.4.tar.gz (12 kB)\n","  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n","  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n","  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","Collecting pyudorandom (from tdigest~=0.5; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading pyudorandom-1.0.0.tar.gz (1.6 kB)\n","  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","Collecting tiledb-cloud>=0.11 (from tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading tiledb_cloud-0.12.29-py3-none-any.whl (789 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m789.5/789.5 kB\u001b[0m \u001b[31m608.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting llvmlite<0.44,>=0.43.0dev0 (from numba>=0.58.0->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading llvmlite-0.43.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (43.9 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m43.9/43.9 MB\u001b[0m \u001b[31m176.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting python-dateutil>=2.8.2 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n","Collecting pytz>=2020.1 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pytz-2024.2-py2.py3-none-any.whl (508 kB)\n","Collecting tzdata>=2022.7 (from pandas->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading tzdata-2024.2-py2.py3-none-any.whl (346 kB)\n","Collecting legacy-api-wrap>=1.4 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading legacy_api_wrap-1.4-py3-none-any.whl (15 kB)\n","Collecting matplotlib>=3.6 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading matplotlib-3.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.3 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.3/8.3 MB\u001b[0m \u001b[31m122.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting patsy (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading patsy-0.5.6-py2.py3-none-any.whl (233 kB)\n","Collecting pynndescent>=0.5 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pynndescent-0.5.13-py3-none-any.whl (56 kB)\n","Collecting seaborn>=0.13 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading seaborn-0.13.2-py3-none-any.whl (294 kB)\n","Collecting session-info (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading session_info-1.0.0.tar.gz (24 kB)\n","  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","Collecting statsmodels>=0.13 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading statsmodels-0.14.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.8 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.8/10.8 MB\u001b[0m \u001b[31m34.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting umap-learn!=0.5.0,>=0.5 (from scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading umap_learn-0.5.6-py3-none-any.whl (85 kB)\n","Collecting pyarrow-hotfix (from somacore==1.0.17->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pyarrow_hotfix-0.6-py3-none-any.whl (7.9 kB)\n","Collecting botocore<1.35.37,>=1.35.16 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading botocore-1.35.36-py3-none-any.whl (12.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.6/12.6 MB\u001b[0m \u001b[31m475.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting wrapt<2.0.0,>=1.10.10 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading wrapt-1.16.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (80 kB)\n","Collecting aioitertools<1.0.0,>=0.5.1 (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading aioitertools-0.12.0-py3-none-any.whl (24 kB)\n","Collecting aiohappyeyeballs>=2.3.0 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading aiohappyeyeballs-2.4.3-py3-none-any.whl (14 kB)\n","Collecting aiosignal>=1.1.2 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading aiosignal-1.3.1-py3-none-any.whl (7.6 kB)\n","Collecting frozenlist>=1.1.1 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading frozenlist-1.4.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (239 kB)\n","Collecting multidict<7.0,>=4.5 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading multidict-6.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (124 kB)\n","Collecting yarl<2.0,>=1.12.0 (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading yarl-1.16.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (318 kB)\n","Collecting async-timeout<5.0,>=4.0; python_version < \"3.11\" (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading async_timeout-4.0.3-py3-none-any.whl (5.7 kB)\n","Collecting MarkupSafe>=2.0 (from jinja2->torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading MarkupSafe-3.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (20 kB)\n","Collecting mpmath<1.4,>=1.1.0 (from sympy==1.13.1; python_version >= \"3.9\"->torch; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading mpmath-1.3.0-py3-none-any.whl (536 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m536.2/536.2 kB\u001b[0m \u001b[31m378.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting cloudpickle<3,>=1.4.1 (from tiledb-cloud>=0.11->tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading cloudpickle-2.2.1-py3-none-any.whl (25 kB)\n","Collecting importlib-metadata (from tiledb-cloud>=0.11->tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading importlib_metadata-8.5.0-py3-none-any.whl (26 kB)\n","Collecting six>=1.10 (from tiledb-cloud>=0.11->tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)\n","Collecting tblib~=1.7 (from tiledb-cloud>=0.11->tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading tblib-1.7.0-py2.py3-none-any.whl (12 kB)\n","Collecting contourpy>=1.0.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading contourpy-1.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (322 kB)\n","Collecting cycler>=0.10 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\n","Collecting fonttools>=4.22.0 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading fonttools-4.54.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.6/4.6 MB\u001b[0m \u001b[31m279.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting kiwisolver>=1.3.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading kiwisolver-1.4.7-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m516.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pillow>=8 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pillow-11.0.0-cp310-cp310-manylinux_2_28_x86_64.whl (4.4 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.4/4.4 MB\u001b[0m \u001b[31m250.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting pyparsing>=2.3.1 (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading pyparsing-3.2.0-py3-none-any.whl (106 kB)\n","Collecting stdlib_list (from session-info->scanpy>=1.9.2->tiledbsoma!=1.14.1,>=1.12.3->cellxgene-census[experimental])\n","  Downloading stdlib_list-0.11.0-py3-none-any.whl (83 kB)\n","Collecting jmespath<2.0.0,>=0.7.1 (from botocore<1.35.37,>=1.35.16->aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading jmespath-1.0.1-py3-none-any.whl (20 kB)\n","Collecting propcache>=0.2.0 (from yarl<2.0,>=1.12.0->aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census[experimental])\n","  Downloading propcache-0.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (208 kB)\n","Collecting zipp>=3.20 (from importlib-metadata->tiledb-cloud>=0.11->tiledb-vector-search~=0.2; extra == \"experimental\"->cellxgene-census[experimental])\n","  Downloading zipp-3.20.2-py3-none-any.whl (9.2 kB)\n","Building wheels for collected packages: accumulation-tree, pyudorandom, session-info\n","  Building wheel for accumulation-tree (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for accumulation-tree: filename=accumulation_tree-0.6.4-cp310-cp310-linux_x86_64.whl size=358618 sha256=f084e2da4a613fa454610aa7cbd85cde82e111fcf161dd275c717894b8d6db7e\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-voivonzy/wheels/8b/66/2c/5b4061508b08a3156d5479641d7dddc69cd84f6e8e20127e80\n","  Building wheel for pyudorandom (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for pyudorandom: filename=pyudorandom-1.0.0-py3-none-any.whl size=2191 sha256=15311215a1e8465f7970920273336212330b69ad7d239172657a422151bbe91b\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-voivonzy/wheels/fc/38/93/20b648196ea752c07f107de4d07748b927989bef8e6effc1ef\n","  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8024 sha256=00e9243877755f63fa260e90417646fb0412e046c615fed93bfe3bc11767f256\n","  Stored in directory: /tmp/pip-ephem-wheel-cache-voivonzy/wheels/6a/aa/b9/eb5d4031476ec10802795b97ccf937b9bd998d68a9b268765a\n","Successfully built accumulation-tree pyudorandom session-info\n","Installing collected packages: array-api-compat, exceptiongroup, numpy, h5py, natsort, packaging, six, python-dateutil, pytz, tzdata, pandas, scipy, anndata, attrs, llvmlite, numba, pyarrow, joblib, legacy-api-wrap, contourpy, cycler, fonttools, kiwisolver, pillow, pyparsing, matplotlib, networkx, patsy, threadpoolctl, scikit-learn, pynndescent, seaborn, stdlib_list, session-info, statsmodels, tqdm, umap-learn, scanpy, pyarrow-hotfix, typing-extensions, somacore, tiledb, tiledbsoma, charset-normalizer, idna, urllib3, certifi, requests, jmespath, botocore, aiohappyeyeballs, frozenlist, aiosignal, multidict, propcache, yarl, async-timeout, aiohttp, wrapt, aioitertools, aiobotocore, fsspec, s3fs, filelock, MarkupSafe, jinja2, nvidia-cuda-nvrtc-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, nvidia-cudnn-cu12, nvidia-cufft-cu12, nvidia-curand-cu12, nvidia-nvjitlink-cu12, nvidia-cusparse-cu12, nvidia-cusolver-cu12, nvidia-nccl-cu12, nvidia-nvtx-cu12, triton, mpmath, sympy, torch, torchdata, scikit-misc, dill, xxhash, multiprocess, pyyaml, huggingface-hub, datasets, accumulation-tree, pyudorandom, tdigest, cloudpickle, zipp, importlib-metadata, tblib, tiledb-cloud, tiledb-vector-search, cellxgene-census\n","  Attempting uninstall: array-api-compat\n","    Found existing installation: array_api_compat 1.9\n","    Uninstalling array_api_compat-1.9:\n","      Successfully uninstalled array_api_compat-1.9\n","  Attempting uninstall: exceptiongroup\n","    Found existing installation: exceptiongroup 1.2.2\n","    Uninstalling exceptiongroup-1.2.2:\n","      Successfully uninstalled exceptiongroup-1.2.2\n","  Attempting uninstall: numpy\n","    Found existing installation: numpy 1.26.4\n","    Uninstalling numpy-1.26.4:\n","      Successfully uninstalled numpy-1.26.4\n","  Attempting uninstall: h5py\n","    Found existing installation: h5py 3.12.1\n","    Uninstalling h5py-3.12.1:\n","      Successfully uninstalled h5py-3.12.1\n","  Attempting uninstall: natsort\n","    Found existing installation: natsort 8.4.0\n","    Uninstalling natsort-8.4.0:\n","      Successfully uninstalled natsort-8.4.0\n","  Attempting uninstall: packaging\n","    Found existing installation: packaging 24.1\n","    Uninstalling packaging-24.1:\n","      Successfully uninstalled packaging-24.1\n","  Attempting uninstall: six\n","    Found existing installation: six 1.16.0\n","    Uninstalling six-1.16.0:\n","      Successfully uninstalled six-1.16.0\n","  Attempting uninstall: python-dateutil\n","    Found existing installation: python-dateutil 2.9.0.post0\n","    Uninstalling python-dateutil-2.9.0.post0:\n","      Successfully uninstalled python-dateutil-2.9.0.post0\n","  Attempting uninstall: pytz\n","    Found existing installation: pytz 2024.2\n","    Uninstalling pytz-2024.2:\n","      Successfully uninstalled pytz-2024.2\n","  Attempting uninstall: tzdata\n","    Found existing installation: tzdata 2024.2\n","    Uninstalling tzdata-2024.2:\n","      Successfully uninstalled tzdata-2024.2\n","  Attempting uninstall: pandas\n","    Found existing installation: pandas 2.2.3\n","    Uninstalling pandas-2.2.3:\n","      Successfully uninstalled pandas-2.2.3\n","  Attempting uninstall: scipy\n","    Found existing installation: scipy 1.14.1\n","    Uninstalling scipy-1.14.1:\n","      Successfully uninstalled scipy-1.14.1\n","  Attempting uninstall: anndata\n","    Found existing installation: anndata 0.10.9\n","    Uninstalling anndata-0.10.9:\n","      Successfully uninstalled anndata-0.10.9\n","  Attempting uninstall: attrs\n","    Found existing installation: attrs 24.2.0\n","    Uninstalling attrs-24.2.0:\n","      Successfully uninstalled attrs-24.2.0\n","  Attempting uninstall: llvmlite\n","    Found existing installation: llvmlite 0.43.0\n","    Uninstalling llvmlite-0.43.0:\n","      Successfully uninstalled llvmlite-0.43.0\n","  Attempting uninstall: numba\n","    Found existing installation: numba 0.60.0\n","    Uninstalling numba-0.60.0:\n","      Successfully uninstalled numba-0.60.0\n","  Attempting uninstall: pyarrow\n","    Found existing installation: pyarrow 17.0.0\n","    Uninstalling pyarrow-17.0.0:\n","      Successfully uninstalled pyarrow-17.0.0\n","  Attempting uninstall: joblib\n","    Found existing installation: joblib 1.4.2\n","    Uninstalling joblib-1.4.2:\n","      Successfully uninstalled joblib-1.4.2\n","  Attempting uninstall: legacy-api-wrap\n","    Found existing installation: legacy-api-wrap 1.4\n","    Uninstalling legacy-api-wrap-1.4:\n","      Successfully uninstalled legacy-api-wrap-1.4\n","  Attempting uninstall: contourpy\n","    Found existing installation: contourpy 1.3.0\n","    Uninstalling contourpy-1.3.0:\n","      Successfully uninstalled contourpy-1.3.0\n","  Attempting uninstall: cycler\n","    Found existing installation: cycler 0.12.1\n","    Uninstalling cycler-0.12.1:\n","      Successfully uninstalled cycler-0.12.1\n","  Attempting uninstall: fonttools\n","    Found existing installation: fonttools 4.54.1\n","    Uninstalling fonttools-4.54.1:\n","      Successfully uninstalled fonttools-4.54.1\n","  Attempting uninstall: kiwisolver\n","    Found existing installation: kiwisolver 1.4.7\n","    Uninstalling kiwisolver-1.4.7:\n","      Successfully uninstalled kiwisolver-1.4.7\n","  Attempting uninstall: pillow\n","    Found existing installation: pillow 11.0.0\n","    Uninstalling pillow-11.0.0:\n","      Successfully uninstalled pillow-11.0.0\n","  Attempting uninstall: pyparsing\n","    Found existing installation: pyparsing 3.2.0\n","    Uninstalling pyparsing-3.2.0:\n","      Successfully uninstalled pyparsing-3.2.0\n","  Attempting uninstall: matplotlib\n","    Found existing installation: matplotlib 3.9.2\n","    Uninstalling matplotlib-3.9.2:\n","      Successfully uninstalled matplotlib-3.9.2\n","  Attempting uninstall: networkx\n","    Found existing installation: networkx 3.4.2\n","    Uninstalling networkx-3.4.2:\n","      Successfully uninstalled networkx-3.4.2\n","  Attempting uninstall: patsy\n","    Found existing installation: patsy 0.5.6\n","    Uninstalling patsy-0.5.6:\n","      Successfully uninstalled patsy-0.5.6\n","  Attempting uninstall: threadpoolctl\n","    Found existing installation: threadpoolctl 3.5.0\n","    Uninstalling threadpoolctl-3.5.0:\n","      Successfully uninstalled threadpoolctl-3.5.0\n","  Attempting uninstall: scikit-learn\n","    Found existing installation: scikit-learn 1.5.2\n","    Uninstalling scikit-learn-1.5.2:\n","      Successfully uninstalled scikit-learn-1.5.2\n","  Attempting uninstall: pynndescent\n","    Found existing installation: pynndescent 0.5.13\n","    Uninstalling pynndescent-0.5.13:\n","      Successfully uninstalled pynndescent-0.5.13\n","  Attempting uninstall: seaborn\n","    Found existing installation: seaborn 0.13.2\n","    Uninstalling seaborn-0.13.2:\n","      Successfully uninstalled seaborn-0.13.2\n","  Attempting uninstall: stdlib_list\n","    Found existing installation: stdlib-list 0.11.0\n","    Uninstalling stdlib-list-0.11.0:\n","      Successfully uninstalled stdlib-list-0.11.0\n","  Attempting uninstall: session-info\n","    Found existing installation: session_info 1.0.0\n","    Uninstalling session_info-1.0.0:\n","      Successfully uninstalled session_info-1.0.0\n","  Attempting uninstall: statsmodels\n","    Found existing installation: statsmodels 0.14.4\n","    Uninstalling statsmodels-0.14.4:\n","      Successfully uninstalled statsmodels-0.14.4\n","  Attempting uninstall: tqdm\n","    Found existing installation: tqdm 4.66.5\n","    Uninstalling tqdm-4.66.5:\n","      Successfully uninstalled tqdm-4.66.5\n","  Attempting uninstall: umap-learn\n","    Found existing installation: umap-learn 0.5.6\n","    Uninstalling umap-learn-0.5.6:\n","      Successfully uninstalled umap-learn-0.5.6\n","  Attempting uninstall: scanpy\n","    Found existing installation: scanpy 1.10.3\n","    Uninstalling scanpy-1.10.3:\n","      Successfully uninstalled scanpy-1.10.3\n","  Attempting uninstall: pyarrow-hotfix\n","    Found existing installation: pyarrow-hotfix 0.6\n","    Uninstalling pyarrow-hotfix-0.6:\n","      Successfully uninstalled pyarrow-hotfix-0.6\n","  Attempting uninstall: typing-extensions\n","    Found existing installation: typing_extensions 4.12.2\n","    Uninstalling typing_extensions-4.12.2:\n","      Successfully uninstalled typing_extensions-4.12.2\n","  Attempting uninstall: somacore\n","    Found existing installation: somacore 1.0.17\n","    Uninstalling somacore-1.0.17:\n","      Successfully uninstalled somacore-1.0.17\n","  Attempting uninstall: tiledb\n","    Found existing installation: tiledb 0.32.3\n","    Uninstalling tiledb-0.32.3:\n","      Successfully uninstalled tiledb-0.32.3\n","  Attempting uninstall: tiledbsoma\n","    Found existing installation: tiledbsoma 1.14.3\n","    Uninstalling tiledbsoma-1.14.3:\n","      Successfully uninstalled tiledbsoma-1.14.3\n","  Attempting uninstall: charset-normalizer\n","    Found existing installation: charset-normalizer 3.4.0\n","    Uninstalling charset-normalizer-3.4.0:\n","      Successfully uninstalled charset-normalizer-3.4.0\n","  Attempting uninstall: idna\n","    Found existing installation: idna 3.10\n","    Uninstalling idna-3.10:\n","      Successfully uninstalled idna-3.10\n","  Attempting uninstall: urllib3\n","    Found existing installation: urllib3 2.2.3\n","    Uninstalling urllib3-2.2.3:\n","      Successfully uninstalled urllib3-2.2.3\n","  Attempting uninstall: certifi\n","    Found existing installation: certifi 2024.8.30\n","    Uninstalling certifi-2024.8.30:\n","      Successfully uninstalled certifi-2024.8.30\n","  Attempting uninstall: requests\n","    Found existing installation: requests 2.32.3\n","    Uninstalling requests-2.32.3:\n","      Successfully uninstalled requests-2.32.3\n","  Attempting uninstall: jmespath\n","    Found existing installation: jmespath 1.0.1\n","    Uninstalling jmespath-1.0.1:\n","      Successfully uninstalled jmespath-1.0.1\n","  Attempting uninstall: botocore\n","    Found existing installation: botocore 1.35.36\n","    Uninstalling botocore-1.35.36:\n","      Successfully uninstalled botocore-1.35.36\n","  Attempting uninstall: aiohappyeyeballs\n","    Found existing installation: aiohappyeyeballs 2.4.3\n","    Uninstalling aiohappyeyeballs-2.4.3:\n","      Successfully uninstalled aiohappyeyeballs-2.4.3\n","  Attempting uninstall: frozenlist\n","    Found existing installation: frozenlist 1.4.1\n","    Uninstalling frozenlist-1.4.1:\n","      Successfully uninstalled frozenlist-1.4.1\n","  Attempting uninstall: aiosignal\n","    Found existing installation: aiosignal 1.3.1\n","    Uninstalling aiosignal-1.3.1:\n","      Successfully uninstalled aiosignal-1.3.1\n","  Attempting uninstall: multidict\n","    Found existing installation: multidict 6.1.0\n","    Uninstalling multidict-6.1.0:\n","      Successfully uninstalled multidict-6.1.0\n","  Attempting uninstall: propcache\n","    Found existing installation: propcache 0.2.0\n","    Uninstalling propcache-0.2.0:\n","      Successfully uninstalled propcache-0.2.0\n","  Attempting uninstall: yarl\n","    Found existing installation: yarl 1.16.0\n","    Uninstalling yarl-1.16.0:\n","      Successfully uninstalled yarl-1.16.0\n","  Attempting uninstall: async-timeout\n","    Found existing installation: async-timeout 4.0.3\n","    Uninstalling async-timeout-4.0.3:\n","      Successfully uninstalled async-timeout-4.0.3\n","  Attempting uninstall: aiohttp\n","    Found existing installation: aiohttp 3.10.10\n","    Uninstalling aiohttp-3.10.10:\n","      Successfully uninstalled aiohttp-3.10.10\n","  Attempting uninstall: wrapt\n","    Found existing installation: wrapt 1.16.0\n","    Uninstalling wrapt-1.16.0:\n","      Successfully uninstalled wrapt-1.16.0\n","  Attempting uninstall: aioitertools\n","    Found existing installation: aioitertools 0.12.0\n","    Uninstalling aioitertools-0.12.0:\n","      Successfully uninstalled aioitertools-0.12.0\n","  Attempting uninstall: aiobotocore\n","    Found existing installation: aiobotocore 2.15.2\n","    Uninstalling aiobotocore-2.15.2:\n","      Successfully uninstalled aiobotocore-2.15.2\n","  Attempting uninstall: fsspec\n","    Found existing installation: fsspec 2024.10.0\n","    Uninstalling fsspec-2024.10.0:\n","      Successfully uninstalled fsspec-2024.10.0\n","  Attempting uninstall: s3fs\n","    Found existing installation: s3fs 2024.10.0\n","    Uninstalling s3fs-2024.10.0:\n","      Successfully uninstalled s3fs-2024.10.0\n","  Attempting uninstall: filelock\n","    Found existing installation: filelock 3.16.1\n","    Uninstalling filelock-3.16.1:\n","      Successfully uninstalled filelock-3.16.1\n","  Attempting uninstall: MarkupSafe\n","    Found existing installation: MarkupSafe 3.0.1\n","    Uninstalling MarkupSafe-3.0.1:\n","      Successfully uninstalled MarkupSafe-3.0.1\n","  Attempting uninstall: jinja2\n","    Found existing installation: Jinja2 3.1.4\n","    Uninstalling Jinja2-3.1.4:\n","      Successfully uninstalled Jinja2-3.1.4\n","  Attempting uninstall: nvidia-cuda-runtime-cu12\n","    Found existing installation: nvidia-cuda-runtime-cu12 12.6.77\n","    Uninstalling nvidia-cuda-runtime-cu12-12.6.77:\n","      Successfully uninstalled nvidia-cuda-runtime-cu12-12.6.77\n","  Attempting uninstall: nvidia-cuda-cupti-cu12\n","    Found existing installation: nvidia-cuda-cupti-cu12 12.6.80\n","    Uninstalling nvidia-cuda-cupti-cu12-12.6.80:\n","      Successfully uninstalled nvidia-cuda-cupti-cu12-12.6.80\n","  Attempting uninstall: nvidia-cublas-cu12\n","    Found existing installation: nvidia-cublas-cu12 12.6.3.3\n","    Uninstalling nvidia-cublas-cu12-12.6.3.3:\n","      Successfully uninstalled nvidia-cublas-cu12-12.6.3.3\n","  Attempting uninstall: nvidia-cudnn-cu12\n","    Found existing installation: nvidia-cudnn-cu12 9.5.0.50\n","    Uninstalling nvidia-cudnn-cu12-9.5.0.50:\n","      Successfully uninstalled nvidia-cudnn-cu12-9.5.0.50\n","  Attempting uninstall: nvidia-cufft-cu12\n","    Found existing installation: nvidia-cufft-cu12 11.3.0.4\n","    Uninstalling nvidia-cufft-cu12-11.3.0.4:\n","      Successfully uninstalled nvidia-cufft-cu12-11.3.0.4\n","  Attempting uninstall: nvidia-curand-cu12\n","    Found existing installation: nvidia-curand-cu12 10.3.7.77\n","    Uninstalling nvidia-curand-cu12-10.3.7.77:\n","      Successfully uninstalled nvidia-curand-cu12-10.3.7.77\n","  Attempting uninstall: nvidia-nvjitlink-cu12\n","    Found existing installation: nvidia-nvjitlink-cu12 12.6.77\n","    Uninstalling nvidia-nvjitlink-cu12-12.6.77:\n","      Successfully uninstalled nvidia-nvjitlink-cu12-12.6.77\n","  Attempting uninstall: nvidia-cusparse-cu12\n","    Found existing installation: nvidia-cusparse-cu12 12.5.4.2\n","    Uninstalling nvidia-cusparse-cu12-12.5.4.2:\n","      Successfully uninstalled nvidia-cusparse-cu12-12.5.4.2\n","  Attempting uninstall: nvidia-cusolver-cu12\n","    Found existing installation: nvidia-cusolver-cu12 11.7.1.2\n","    Uninstalling nvidia-cusolver-cu12-11.7.1.2:\n","      Successfully uninstalled nvidia-cusolver-cu12-11.7.1.2\n","  Attempting uninstall: nvidia-nccl-cu12\n","    Found existing installation: nvidia-nccl-cu12 2.23.4\n","    Uninstalling nvidia-nccl-cu12-2.23.4:\n","      Successfully uninstalled nvidia-nccl-cu12-2.23.4\n","  Attempting uninstall: mpmath\n","    Found existing installation: mpmath 1.3.0\n","    Uninstalling mpmath-1.3.0:\n","      Successfully uninstalled mpmath-1.3.0\n","  Attempting uninstall: sympy\n","    Found existing installation: sympy 1.13.3\n","    Uninstalling sympy-1.13.3:\n","      Successfully uninstalled sympy-1.13.3\n","  Attempting uninstall: torch\n","    Found existing installation: torch 2.4.1+cu121\n","    Uninstalling torch-2.4.1+cu121:\n","      Successfully uninstalled torch-2.4.1+cu121\n","  Attempting uninstall: torchdata\n","    Found existing installation: torchdata 0.9.0\n","    Uninstalling torchdata-0.9.0:\n","      Successfully uninstalled torchdata-0.9.0\n","  Attempting uninstall: scikit-misc\n","    Found existing installation: scikit-misc 0.3.1\n","    Uninstalling scikit-misc-0.3.1:\n","      Successfully uninstalled scikit-misc-0.3.1\n","  Attempting uninstall: dill\n","    Found existing installation: dill 0.3.8\n","    Uninstalling dill-0.3.8:\n","      Successfully uninstalled dill-0.3.8\n","  Attempting uninstall: xxhash\n","    Found existing installation: xxhash 3.5.0\n","    Uninstalling xxhash-3.5.0:\n","      Successfully uninstalled xxhash-3.5.0\n","  Attempting uninstall: multiprocess\n","    Found existing installation: multiprocess 0.70.17\n","    Uninstalling multiprocess-0.70.17:\n","      Successfully uninstalled multiprocess-0.70.17\n","  Attempting uninstall: pyyaml\n","    Found existing installation: PyYAML 6.0.2\n","    Uninstalling PyYAML-6.0.2:\n","      Successfully uninstalled PyYAML-6.0.2\n","  Attempting uninstall: huggingface-hub\n","    Found existing installation: huggingface-hub 0.24.7\n","    Uninstalling huggingface-hub-0.24.7:\n","      Successfully uninstalled huggingface-hub-0.24.7\n","  Attempting uninstall: datasets\n","    Found existing installation: datasets 2.21.0\n","    Uninstalling datasets-2.21.0:\n","      Successfully uninstalled datasets-2.21.0\n","  Attempting uninstall: accumulation-tree\n","    Found existing installation: accumulation_tree 0.6.4\n","    Uninstalling accumulation_tree-0.6.4:\n","      Successfully uninstalled accumulation_tree-0.6.4\n","  Attempting uninstall: pyudorandom\n","    Found existing installation: pyudorandom 1.0.0\n","    Uninstalling pyudorandom-1.0.0:\n","      Successfully uninstalled pyudorandom-1.0.0\n","  Attempting uninstall: tdigest\n","    Found existing installation: tdigest 0.5.2.2\n","    Uninstalling tdigest-0.5.2.2:\n","      Successfully uninstalled tdigest-0.5.2.2\n","  Attempting uninstall: cloudpickle\n","    Found existing installation: cloudpickle 2.2.1\n","    Uninstalling cloudpickle-2.2.1:\n","      Successfully uninstalled cloudpickle-2.2.1\n","  Attempting uninstall: zipp\n","    Found existing installation: zipp 3.20.2\n","    Uninstalling zipp-3.20.2:\n","      Successfully uninstalled zipp-3.20.2\n","  Attempting uninstall: importlib-metadata\n","    Found existing installation: importlib_metadata 8.5.0\n","    Uninstalling importlib_metadata-8.5.0:\n","      Successfully uninstalled importlib_metadata-8.5.0\n","  Attempting uninstall: tblib\n","    Found existing installation: tblib 1.7.0\n","    Uninstalling tblib-1.7.0:\n","      Successfully uninstalled tblib-1.7.0\n","  Attempting uninstall: tiledb-cloud\n","    Found existing installation: tiledb-cloud 0.12.29\n","    Uninstalling tiledb-cloud-0.12.29:\n","      Successfully uninstalled tiledb-cloud-0.12.29\n","  Attempting uninstall: tiledb-vector-search\n","    Found existing installation: tiledb-vector-search 0.10.2\n","    Uninstalling tiledb-vector-search-0.10.2:\n","      Successfully uninstalled tiledb-vector-search-0.10.2\n","  Attempting uninstall: cellxgene-census\n","    Found existing installation: cellxgene-census 1.16.2\n","    Uninstalling cellxgene-census-1.16.2:\n","      Successfully uninstalled cellxgene-census-1.16.2\n","\u001b[31mERROR: pip's legacy dependency resolver does not consider dependency conflicts when selecting packages. This behaviour is the source of the following dependency conflicts.\n","cudf-cu12 24.10.1 requires pandas<2.2.3dev0,>=2.0, but you'll have pandas 2.2.3 which is incompatible.\n","dask 2024.8.2 requires cloudpickle>=3.0.0, but you'll have cloudpickle 2.2.1 which is incompatible.\n","datasets 2.21.0 requires fsspec[http]<=2024.6.1,>=2023.1.0, but you'll have fsspec 2024.10.0 which is incompatible.\n","fastai 2.7.17 requires torch<2.5,>=1.10, but you'll have torch 2.5.0 which is incompatible.\n","gcsfs 2024.6.1 requires fsspec==2024.6.1, but you'll have fsspec 2024.10.0 which is incompatible.\n","gensim 4.3.3 requires scipy<1.14.0,>=1.7.0, but you'll have scipy 1.14.1 which is incompatible.\n","google-colab 1.0.0 requires pandas==2.2.2, but you'll have pandas 2.2.3 which is incompatible.\n","multiprocess 0.70.17 requires dill>=0.3.9, but you'll have dill 0.3.8 which is incompatible.\n","torchaudio 2.4.1+cu121 requires torch==2.4.1, but you'll have torch 2.5.0 which is incompatible.\n","torchvision 0.19.1+cu121 requires torch==2.4.1, but you'll have torch 2.5.0 which is incompatible.\u001b[0m\u001b[31m\n","\u001b[0mSuccessfully installed MarkupSafe-3.0.2 accumulation-tree-0.6.4 aiobotocore-2.15.2 aiohappyeyeballs-2.4.3 aiohttp-3.10.10 aioitertools-0.12.0 aiosignal-1.3.1 anndata-0.10.9 array-api-compat-1.9 async-timeout-4.0.3 attrs-24.2.0 botocore-1.35.36 cellxgene-census-1.16.2 certifi-2024.8.30 charset-normalizer-3.4.0 cloudpickle-2.2.1 contourpy-1.3.0 cycler-0.12.1 datasets-2.21.0 dill-0.3.8 exceptiongroup-1.2.2 filelock-3.16.1 fonttools-4.54.1 frozenlist-1.4.1 fsspec-2024.10.0 h5py-3.12.1 huggingface-hub-0.26.1 idna-3.10 importlib-metadata-8.5.0 jinja2-3.1.4 jmespath-1.0.1 joblib-1.4.2 kiwisolver-1.4.7 legacy-api-wrap-1.4 llvmlite-0.43.0 matplotlib-3.9.2 mpmath-1.3.0 multidict-6.1.0 multiprocess-0.70.17 natsort-8.4.0 networkx-3.4.2 numba-0.60.0 numpy-1.26.4 nvidia-cublas-cu12-12.4.5.8 nvidia-cuda-cupti-cu12-12.4.127 nvidia-cuda-nvrtc-cu12-12.4.127 nvidia-cuda-runtime-cu12-12.4.127 nvidia-cudnn-cu12-9.1.0.70 nvidia-cufft-cu12-11.2.1.3 nvidia-curand-cu12-10.3.5.147 nvidia-cusolver-cu12-11.6.1.9 nvidia-cusparse-cu12-12.3.1.170 nvidia-nccl-cu12-2.21.5 nvidia-nvjitlink-cu12-12.4.127 nvidia-nvtx-cu12-12.4.127 packaging-24.1 pandas-2.2.3 patsy-0.5.6 pillow-11.0.0 propcache-0.2.0 pyarrow-17.0.0 pyarrow-hotfix-0.6 pynndescent-0.5.13 pyparsing-3.2.0 python-dateutil-2.9.0.post0 pytz-2024.2 pyudorandom-1.0.0 pyyaml-6.0.2 requests-2.32.3 s3fs-2024.10.0 scanpy-1.10.3 scikit-learn-1.5.2 scikit-misc-0.3.1 scipy-1.14.1 seaborn-0.13.2 session-info-1.0.0 six-1.16.0 somacore-1.0.17 statsmodels-0.14.4 stdlib_list-0.11.0 sympy-1.13.1 tblib-1.7.0 tdigest-0.5.2.2 threadpoolctl-3.5.0 tiledb-0.32.3 tiledb-cloud-0.12.29 tiledb-vector-search-0.10.2 tiledbsoma-1.14.3 torch-2.5.0 torchdata-0.9.0 tqdm-4.66.5 triton-3.1.0 typing-extensions-4.12.2 tzdata-2024.2 umap-learn-0.5.6 urllib3-2.2.3 wrapt-1.16.0 xxhash-3.5.0 yarl-1.16.0 zipp-3.20.2\n"]},{"output_type":"display_data","data":{"application/vnd.colab-display-data+json":{"pip_warning":{"packages":["PIL","certifi","cycler","dateutil","kiwisolver","six"]},"id":"3e183bd0c65f4026bb6bd683ebb506c5"}},"metadata":{}}]},{"cell_type":"markdown","source":["# **Step 2: Python Quick Start**"],"metadata":{"id":"LpSo2Ks5lkix"}},{"cell_type":"code","source":["import cellxgene_census\n","\n","help(cellxgene_census)\n","help(cellxgene_census.get_anndata)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pjgw8MlDlo8R","executionInfo":{"status":"ok","timestamp":1729573080277,"user_tz":480,"elapsed":9547,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"514d71c3-20ca-44ca-9585-97d9ab207fbb"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/tiledb/cloud/config.py:96: UserWarning: You must first login before you can run commands. Please run tiledb.cloud.login.\n","  warnings.warn(\n"]},{"output_type":"stream","name":"stdout","text":["Help on package cellxgene_census:\n","\n","NAME\n","    cellxgene_census - An API to facilitate use of the CZI Science CELLxGENE Census. The Census is a versioned container of single-cell data hosted at `CELLxGENE Discover`_.\n","\n","DESCRIPTION\n","    The API is built on the `tiledbsoma` SOMA API, and provides a number of helper functions including:\n","    \n","        * Open a named version of the Census, for use with the SOMA API\n","        * Get a list of available Census versions, and for each version, a description\n","        * Get a slice of the Census as an AnnData, for use with ScanPy\n","        * Get the URI for, or directly download, underlying data in H5AD format\n","    \n","    For more information on the API, visit the `cellxgene_census repo`_. For more information on SOMA, see the `tiledbsoma repo`_.\n","    \n","    .. _CELLxGENE Discover:\n","        https://cellxgene.cziscience.com/\n","    \n","    .. _cellxgene_census repo:\n","        https://github.com/chanzuckerberg/cellxgene-census/\n","    \n","    .. _tiledbsoma repo:\n","        https://github.com/single-cell-data/TileDB-SOMA\n","\n","PACKAGE CONTENTS\n","    _experiment\n","    _get_anndata\n","    _open\n","    _presence_matrix\n","    _release_directory\n","    _testing (package)\n","    _util\n","    experimental (package)\n","\n","FUNCTIONS\n","    download_source_h5ad(dataset_id: str, to_path: str, *, census_version: str = 'stable', progress_bar: bool = True) -> None\n","        Download the source H5AD dataset, for the given `dataset_id`, to the user-specified\n","        file name.\n","        \n","        Args:\n","            dataset_id\n","                Fetch the source (original) H5AD associated with this ``dataset_id``.\n","            to_path:\n","                The file name where the downloaded H5AD will be written. Must not already exist.\n","            census_version:\n","                The census version name. Defaults to ``\"stable\"``.\n","            progress_bar:\n","                Whether to display a progress bar. Defaults to ``True``.\n","        \n","        Raises:\n","            ValueError: if the path already exists (i.e., will not overwrite an existing file), or is not a file.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        See Also:\n","            :func:`get_source_h5ad_uri`: Look up the location of the source H5AD.\n","        \n","        Examples:\n","            >>> download_source_h5ad(\"8e47ed12-c658-4252-b126-381df8d52a3d\", to_path=\"/tmp/data.h5ad\")\n","    \n","    get_anndata(census: tiledbsoma._collection.Collection, organism: str, measurement_name: str = 'RNA', X_name: str = 'raw', X_layers: collections.abc.Sequence[str] | None = (), obsm_layers: collections.abc.Sequence[str] | None = (), obsp_layers: collections.abc.Sequence[str] | None = (), varm_layers: collections.abc.Sequence[str] | None = (), varp_layers: collections.abc.Sequence[str] | None = (), obs_value_filter: str | None = None, obs_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = None, var_value_filter: str | None = None, var_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = None, column_names: somacore.query.query.AxisColumnNames | None = None, obs_embeddings: collections.abc.Sequence[str] | None = (), var_embeddings: collections.abc.Sequence[str] | None = (), obs_column_names: collections.abc.Sequence[str] | None = None, var_column_names: collections.abc.Sequence[str] | None = None) -> anndata._core.anndata.AnnData\n","        Convenience wrapper around :class:`tiledbsoma.Experiment` query, to build and execute a query,\n","        and return it as an :class:`anndata.AnnData` object.\n","        \n","        Args:\n","            census:\n","                The census object, usually returned by :func:`open_soma`.\n","            organism:\n","                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``.\n","            measurement_name:\n","                The measurement object to query. Defaults to ``\"RNA\"``.\n","            X_name:\n","                The ``X`` layer to query. Defaults to ``\"raw\"``.\n","            X_layers:\n","                Additional layers to add to :attr:`anndata.AnnData.layers`.\n","            obs_value_filter:\n","                Value filter for the ``obs`` metadata. Value is a filter query written in the\n","                SOMA ``value_filter`` syntax.\n","            obs_coords:\n","                Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n","                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","            var_value_filter:\n","                Value filter for the ``var`` metadata. Value is a filter query written in the\n","                SOMA ``value_filter`` syntax.\n","            var_coords:\n","                Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n","                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","            obsm_layers:\n","                Additional obsm layers to read and return in the ``obsm`` slot.\n","            obsp_layers:\n","                Additional obsp layers to read and return in the ``obsp`` slot.\n","            varm_layers:\n","                Additional varm layers to read and return in the ``varm`` slot.\n","            varp_layers:\n","                Additional varp layers to read and return in the ``varp`` slot.\n","            obs_embeddings:\n","                Additional embeddings to be returned as part of the ``obsm`` slot.\n","                Use :func:`get_all_available_embeddings` to retrieve available embeddings\n","                for this Census version and organism.\n","            var_embeddings:\n","                Additional embeddings to be returned as part of the ``varm`` slot.\n","                Use :func:`get_all_available_embeddings` to retrieve available embeddings\n","                for this Census version and organism.\n","            obs_column_names:\n","                Columns to fetch for ``obs`` dataframe.\n","            var_column_names:\n","                Columns to fetch for ``var`` dataframe.\n","        \n","        Returns:\n","            An :class:`anndata.AnnData` object containing the census slice.\n","        \n","        Lifecycle:\n","            experimental\n","        \n","        Examples:\n","            >>> get_anndata(census, \"Mus musculus\", obs_value_filter=\"tissue_general in ['brain', 'lung']\")\n","        \n","            >>> get_anndata(census, \"Homo sapiens\", obs_column_names=[\"tissue\"])\n","        \n","            >>> get_anndata(census, \"Homo sapiens\", obs_coords=slice(0, 1000))\n","    \n","    get_census_mirror_directory() -> dict[str, cellxgene_census._release_directory.CensusMirror]\n","        Get the directory of Census mirrors currently available.\n","        \n","        Returns:\n","            A dictionary that contains mirror names and their corresponding info,\n","            like the provider and the region.\n","        \n","        Lifecycle:\n","            maturing\n","    \n","    get_census_version_description(census_version: str) -> cellxgene_census._release_directory.CensusVersionDescription\n","        Get release description for given Census version, from the Census release directory.\n","        \n","        Args:\n","            census_version:\n","                The census version name.\n","        \n","        Returns:\n","            ``CensusVersionDescription`` - a dictionary containing a description of the release.\n","        \n","        Raises:\n","            ValueError: if unknown ``census_version`` value.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        See Also:\n","            :func:`get_census_version_directory`: returns the entire directory as a dict.\n","        \n","        Examples:\n","            >>> cellxgene_census.get_census_version_description(\"latest\")\n","            {'release_date': None,\n","            'release_build': '2022-12-01',\n","            'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n","            's3_region': 'us-west-2'},\n","            'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n","            's3_region': 'us-west-2'}}\n","    \n","    get_census_version_directory(*, lts: bool | None = None, retracted: bool | None = False) -> dict[str, cellxgene_census._release_directory.CensusVersionDescription]\n","        Get the directory of Census versions currently available, optionally filtering by specified\n","        flags. If a filtering flag is not specified, Census versions will not be filtered by that flag.\n","        Defaults to including both \"long-term stable\" (LTS) and weekly Census versions, and excluding\n","        retracted versions.\n","        \n","        Args:\n","            lts:\n","                A filtering flag to either include or exclude long-term stable releases in the result.\n","                If ``None``, no filtering is performed based on this flag. Defaults to ``None``, which\n","                includes both LTS and non-LTS (weekly) versions.\n","            retracted:\n","                A filtering flag to either include or exclude retracted releases in the result. If\n","                ``None``, no filtering is performed based on this flag. Defaults to ``False``, which\n","                excludes retracted releases in the result.\n","        \n","        Returns:\n","            A dictionary that contains Census version names and their corresponding descriptions. Census\n","            versions are always named by their release date (``YYYY-MM-DD``) but may also have aliases.\n","            If an alias is specified, the Census version will appear multiple times in the dictionary,\n","            once under it's release date name, and again for each alias. Aliases may be: ``\"stable\"``,\n","            ``\"latest\"``, or ``\"V#\"``. The ``\"stable\"`` alias is used for the most recent LTS release,\n","            the ``\"latest\"`` alias is used for the most recent weekly release, and the ``\"V#\"`` aliases\n","            are used to identify LTS releases by a sequentially incrementing version number.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        See Also:\n","            :func:`get_census_version_description`: get description by ``census_version``.\n","        \n","        Examples:\n","            Get all LTS and weekly versions, but exclude retracted LTS versions:\n","        \n","            >>> cellxgene_census.get_census_version_directory()\n","                {\n","                    'stable': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    },\n","                    'latest': {\n","                        'release_date': None,\n","                        'release_build': '2022-12-01',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    },\n","                    'V2': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    },\n","                    '2022-12-01': {\n","                        'release_date': None,\n","                        'release_build': '2022-12-01',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': False, 'retracted': False}\n","                    },\n","                    '2022-11-29': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    }\n","                }\n","        \n","            Get only LTS versions that are not retracted:\n","        \n","            >>> cellxgene_census.get_census_version_directory(lts=True)\n","                {\n","                    'stable': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    },\n","                    'V2': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    },\n","                    '2022-11-29': {\n","                        'release_date': None,\n","                        'release_build': '2022-11-29',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': False}\n","                    }\n","                }\n","        \n","            Get only retracted releases:\n","        \n","            >>> cellxgene_census.get_census_version_directory(retracted=True)\n","                {\n","                    'V1': {\n","                        'release_date': None,\n","                        'release_build': '2022-10-15',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': True},\n","                        'retraction': {\n","                            'date': '2022-10-30',\n","                            'reason': 'mistakes happen',\n","                            'info_url': 'http://cellxgene.com/census/errata/v1',\n","                            'replaced_by': 'V2'\n","                        },\n","                    },\n","                    '2022-10-15': {\n","                        'release_date': None,\n","                        'release_build': '2022-10-15',\n","                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/soma/',\n","                                 's3_region': 'us-west-2'},\n","                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/h5ads/',\n","                                  's3_region': 'us-west-2'},\n","                        'flags': {'lts': True, 'retracted': True},\n","                        'retraction': {\n","                            'date': '2022-10-30',\n","                            'reason': 'mistakes happen',\n","                            'info_url': 'http://cellxgene.com/census/errata/v1',\n","                            'replaced_by': 'V2'\n","                        }\n","                    }\n","                }\n","    \n","    get_default_soma_context(tiledb_config: dict[str, typing.Any] | None = None) -> tiledbsoma.options._soma_tiledb_context.SOMATileDBContext\n","        Return a :class:`tiledbsoma.SOMATileDBContext` with sensible defaults that can be further customized by the\n","        user. The customized context can then be passed to :func:`cellxgene_census.open_soma` with the ``context``\n","        argument or to :meth:`somacore.SOMAObject.open` with the ``context`` argument, such as\n","        :meth:`tiledbsoma.Experiment.open`. Use the :meth:`tiledbsoma.SOMATileDBContext.replace` method on the returned\n","        object to customize its settings further.\n","        \n","        Args:\n","            tiledb_config:\n","                A dictionary of TileDB configuration parameters. If specified, the parameters will override the\n","                defaults. If not specified, the default configuration will be returned.\n","        \n","        Returns:\n","            A :class:`tiledbsoma.SOMATileDBContext` object with sensible defaults.\n","        \n","        Examples:\n","            To reduce the amount of memory used by TileDB-SOMA I/O operations:\n","        \n","            .. highlight:: python\n","            .. code-block:: python\n","        \n","                ctx = cellxgene_census.get_default_soma_context(\n","                    tiledb_config={\"py.init_buffer_bytes\": 128 * 1024**2,\n","                                   \"soma.init_buffer_bytes\": 128 * 1024**2})\n","                c = census.open_soma(uri=\"s3://my-private-bucket/census/soma\", context=ctx)\n","        \n","            To access a copy of the Census located in a private bucket that is located in a different S3 region, use:\n","        \n","            .. highlight:: python\n","            .. code-block:: python\n","        \n","                ctx = cellxgene_census.get_default_soma_context(\n","                    tiledb_config={\"vfs.s3.no_sign_request\": \"false\",\n","                                   \"vfs.s3.region\": \"us-east-1\"})\n","                c = census.open_soma(uri=\"s3://my-private-bucket/census/soma\", context=ctx)\n","        \n","        Lifecycle:\n","            experimental\n","    \n","    get_obs(census: tiledbsoma._collection.Collection, organism: str, *, value_filter: str | None = None, coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = slice(None, None, None), column_names: collections.abc.Sequence[str] | None = None) -> pandas.core.frame.DataFrame\n","        Get the observation metadata for a query on the census.\n","        \n","        Args:\n","            census:\n","                The census object, usually returned by :func:`open_soma`.\n","            organism:\n","                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``\n","            value_filter:\n","                Value filter for the ``obs`` metadata. Value is a filter query written in the\n","                SOMA ``value_filter`` syntax.\n","            coords:\n","                Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n","                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","            column_names:\n","                Columns to fetch.\n","        \n","        Returns:\n","            A :class:`pandas.DataFrame` object containing metadata for the queried slice.\n","    \n","    get_presence_matrix(census: tiledbsoma._collection.Collection, organism: str, measurement_name: str = 'RNA') -> scipy.sparse._csr.csr_matrix\n","        Read the feature dataset presence matrix and return as a :class:`scipy.sparse.csr_array`. The\n","        returned sparse matrix is indexed on the first dimension by the dataset ``soma_joinid`` values,\n","        and on the second dimension by the ``var`` :class:`pandas.DataFrame` ``soma_joinid`` values.\n","        \n","        Args:\n","            census:\n","                The census from which to read the presence matrix.\n","            organism:\n","                The organism to query, usually one of ``\"Homo sapiens\"`` or ``\"Mus musculus\"``.\n","            measurement_name:\n","                The measurement object to query. Deafults to ``\"RNA\"``.\n","        \n","        Returns:\n","            A :class:`scipy.sparse.csr_array` object containing the presence matrix.\n","        \n","        Raises:\n","            ValueError: if the organism cannot be found.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        Examples:\n","            >>> get_presence_matrix(census, \"Homo sapiens\", \"RNA\")\n","            <321x60554 sparse array of type '<class 'numpy.uint8'>'\n","            with 6441269 stored elements in Compressed Sparse Row format>\n","    \n","    get_source_h5ad_uri(dataset_id: str, *, census_version: str = 'stable') -> cellxgene_census._release_directory.CensusLocator\n","        Open the named version of the census, and return the URI for the ``dataset_id``. This\n","        does not guarantee that the H5AD exists or is accessible to the user.\n","        \n","        Args:\n","            dataset_id:\n","                The ``dataset_id`` of interest.\n","            census_version:\n","                The census version. Defaults to ``\"stable\"``.\n","        \n","        Returns:\n","            A :class:`cellxgene_census._release_directory.CensusLocator` object that contains the URI and optional S3 region\n","            for the source H5AD.\n","        \n","        Raises:\n","            KeyError: if either ``dataset_id`` or ``census_version`` do not exist.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        Examples:\n","            >>> cellxgene_census.get_source_h5ad_uri(\"cb5efdb0-f91c-4cbd-9ad4-9d4fa41c572d\")\n","            {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/cb5efdb0-f91c-4cbd-9ad4-9d4fa41c572d.h5ad',\n","            's3_region': 'us-west-2'}\n","    \n","    get_var(census: tiledbsoma._collection.Collection, organism: str, *, value_filter: str | None = None, coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = slice(None, None, None), column_names: collections.abc.Sequence[str] | None = None) -> pandas.core.frame.DataFrame\n","        Get the variable metadata for a query on the census.\n","        \n","        Args:\n","            census:\n","                The census object, usually returned by :func:`open_soma`.\n","            organism:\n","                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``\n","            value_filter:\n","                Value filter for the ``var`` metadata. Value is a filter query written in the\n","                SOMA ``value_filter`` syntax.\n","            coords:\n","                Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n","                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","            column_names:\n","                Columns to fetch.\n","        \n","        Returns:\n","            A :class:`pandas.DataFrame` object containing metadata for the queried slice.\n","    \n","    open_soma(*, census_version: str | None = 'stable', mirror: str | None = None, uri: str | None = None, tiledb_config: dict[str, typing.Any] | None = None, context: tiledbsoma.options._soma_tiledb_context.SOMATileDBContext | None = None) -> tiledbsoma._collection.Collection\n","        Open the Census by version or URI.\n","        \n","        Args:\n","            census_version:\n","                The version of the Census, e.g. ``\"latest\"`` or ``\"stable\"``. Defaults to ``\"stable\"``.\n","            mirror:\n","                The mirror used to retrieve the Census. If not specified, a suitable mirror\n","                will be chosen automatically.\n","            uri:\n","                The URI containing the Census SOMA objects. If specified, will take precedence\n","                over ``census_version`` parameter.\n","            tiledb_config:\n","                A dictionary of TileDB configuration parameters that will be used to open the SOMA object. Optional,\n","                defaults to ``None``. If specified, the parameters will override the default settings specified by\n","                ``get_default_soma_context().tiledb_config``. Only one of the ``tiledb_config`` and ``context`` params\n","                can be specified.\n","            context:\n","                A custom :class:`tiledbsoma.SOMATileDBContext` that will be used to open the SOMA object.\n","                Optional, defaults to ``None``. Only one of the ``tiledb_config`` and ``context`` params can be specified.\n","        \n","        Returns:\n","            A :class:`tiledbsoma.Collection` object containing the top-level census.\n","            It can be used as a context manager, which will automatically close upon exit.\n","        \n","        Raises:\n","            ValueError: if the census cannot be found, the URI cannot be opened, neither a URI\n","                or a version are specified, or an invalid mirror is provided.\n","        \n","        See Also:\n","            - :func:`get_source_h5ad_uri`: Look up the location of the source H5AD.\n","            - :func:`get_default_soma_context`: Get a default SOMA context that can be updated with custom settings and used\n","              as the ``context`` argument.\n","        \n","        Lifecycle:\n","            maturing\n","        \n","        Examples:\n","            Open the default Census version, using a context manager which will automatically\n","            close the Census upon exit of the context.\n","        \n","            >>> with cellxgene_census.open_soma() as census:\n","                    ...\n","        \n","            Open and close:\n","        \n","            >>> census = cellxgene_census.open_soma()\n","                ...\n","                census.close()\n","        \n","            Open a specific Census by version:\n","        \n","            >>> with cellxgene_census.open_soma(\"2022-12-31\") as census:\n","                    ...\n","        \n","            Open a Census by S3 URI, rather than by version.\n","        \n","            >>> with cellxgene_census.open_soma(uri=\"s3://bucket/path\") as census:\n","                    ...\n","        \n","            Open a Census by path (file:// URI), rather than by version.\n","        \n","            >>> with cellxgene_census.open_soma(uri=\"/tmp/census\") as census:\n","                    ...\n","        \n","            Open a Census using a mirror.\n","        \n","            >>> with cellxgene_census.open_soma(mirror=\"s3-us-west-2\") as census:\n","                    ...\n","        \n","            Open a Census with a custom TileDB configuration setting.\n","        \n","            >>> with cellxgene_census.open_soma(tiledb_config={\"py.init_buffer_bytes\": 128 * 1024**2}) as census:\n","                    ...\n","\n","DATA\n","    __all__ = ['download_source_h5ad', 'get_anndata', 'get_obs', 'get_var'...\n","\n","VERSION\n","    1.16.2\n","\n","FILE\n","    /usr/local/lib/python3.10/dist-packages/cellxgene_census/__init__.py\n","\n","\n","Help on function get_anndata in module cellxgene_census._get_anndata:\n","\n","get_anndata(census: tiledbsoma._collection.Collection, organism: str, measurement_name: str = 'RNA', X_name: str = 'raw', X_layers: collections.abc.Sequence[str] | None = (), obsm_layers: collections.abc.Sequence[str] | None = (), obsp_layers: collections.abc.Sequence[str] | None = (), varm_layers: collections.abc.Sequence[str] | None = (), varp_layers: collections.abc.Sequence[str] | None = (), obs_value_filter: str | None = None, obs_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = None, var_value_filter: str | None = None, var_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[Any, numpy.dtype[numpy.integer]], numpy.ndarray[Any, numpy.dtype[numpy.datetime64]]] = None, column_names: somacore.query.query.AxisColumnNames | None = None, obs_embeddings: collections.abc.Sequence[str] | None = (), var_embeddings: collections.abc.Sequence[str] | None = (), obs_column_names: collections.abc.Sequence[str] | None = None, var_column_names: collections.abc.Sequence[str] | None = None) -> anndata._core.anndata.AnnData\n","    Convenience wrapper around :class:`tiledbsoma.Experiment` query, to build and execute a query,\n","    and return it as an :class:`anndata.AnnData` object.\n","    \n","    Args:\n","        census:\n","            The census object, usually returned by :func:`open_soma`.\n","        organism:\n","            The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``.\n","        measurement_name:\n","            The measurement object to query. Defaults to ``\"RNA\"``.\n","        X_name:\n","            The ``X`` layer to query. Defaults to ``\"raw\"``.\n","        X_layers:\n","            Additional layers to add to :attr:`anndata.AnnData.layers`.\n","        obs_value_filter:\n","            Value filter for the ``obs`` metadata. Value is a filter query written in the\n","            SOMA ``value_filter`` syntax.\n","        obs_coords:\n","            Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n","            May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","        var_value_filter:\n","            Value filter for the ``var`` metadata. Value is a filter query written in the\n","            SOMA ``value_filter`` syntax.\n","        var_coords:\n","            Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n","            May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n","        obsm_layers:\n","            Additional obsm layers to read and return in the ``obsm`` slot.\n","        obsp_layers:\n","            Additional obsp layers to read and return in the ``obsp`` slot.\n","        varm_layers:\n","            Additional varm layers to read and return in the ``varm`` slot.\n","        varp_layers:\n","            Additional varp layers to read and return in the ``varp`` slot.\n","        obs_embeddings:\n","            Additional embeddings to be returned as part of the ``obsm`` slot.\n","            Use :func:`get_all_available_embeddings` to retrieve available embeddings\n","            for this Census version and organism.\n","        var_embeddings:\n","            Additional embeddings to be returned as part of the ``varm`` slot.\n","            Use :func:`get_all_available_embeddings` to retrieve available embeddings\n","            for this Census version and organism.\n","        obs_column_names:\n","            Columns to fetch for ``obs`` dataframe.\n","        var_column_names:\n","            Columns to fetch for ``var`` dataframe.\n","    \n","    Returns:\n","        An :class:`anndata.AnnData` object containing the census slice.\n","    \n","    Lifecycle:\n","        experimental\n","    \n","    Examples:\n","        >>> get_anndata(census, \"Mus musculus\", obs_value_filter=\"tissue_general in ['brain', 'lung']\")\n","    \n","        >>> get_anndata(census, \"Homo sapiens\", obs_column_names=[\"tissue\"])\n","    \n","        >>> get_anndata(census, \"Homo sapiens\", obs_coords=slice(0, 1000))\n","\n"]}]},{"cell_type":"markdown","source":["**Querying slice of cell metadata**"],"metadata":{"id":"ewG8nVWTl6w_"}},{"cell_type":"code","source":["import cellxgene_census\n","\n","with cellxgene_census.open_soma() as census:\n","\n","    # Reads SOMADataFrame as a slice\n","    cell_metadata = census[\"census_data\"][\"homo_sapiens\"].obs.read(\n","        value_filter = \"sex == 'female' and cell_type in ['microglial cell', 'neuron']\",\n","        column_names = [\"assay\", \"cell_type\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]\n","    )\n","\n","    # Concatenates results to pyarrow.Table\n","    cell_metadata = cell_metadata.concat()\n","\n","    # Converts to pandas.DataFrame\n","    cell_metadata = cell_metadata.to_pandas()\n","\n","    print(cell_metadata)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SJSagSE2mAXU","executionInfo":{"status":"ok","timestamp":1729573198073,"user_tz":480,"elapsed":8310,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"9e0b5c27-c65e-4b19-f0ef-79903ba71958"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n","INFO:cellxgene_census:The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n"]},{"output_type":"stream","name":"stdout","text":["            assay        cell_type                          tissue  \\\n","0       10x 5' v1           neuron                            lung   \n","1       10x 5' v1           neuron                            lung   \n","2       10x 5' v1           neuron                            lung   \n","3       10x 5' v1           neuron                            lung   \n","4       10x 5' v1           neuron                            lung   \n","...           ...              ...                             ...   \n","732776  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n","732777  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n","732778  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n","732779  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n","732780  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n","\n","       tissue_general suspension_type   disease     sex  \n","0                lung            cell    normal  female  \n","1                lung            cell    normal  female  \n","2                lung            cell    normal  female  \n","3                lung            cell    normal  female  \n","4                lung            cell    normal  female  \n","...               ...             ...       ...     ...  \n","732776          brain         nucleus    normal  female  \n","732777          brain         nucleus    normal  female  \n","732778          brain         nucleus  dementia  female  \n","732779          brain         nucleus  dementia  female  \n","732780          brain         nucleus  dementia  female  \n","\n","[732781 rows x 7 columns]\n"]}]},{"cell_type":"markdown","source":["**Obtaining a slice as AnnData**"],"metadata":{"id":"7NsdXglrmOSE"}},{"cell_type":"code","source":["import cellxgene_census\n","\n","with cellxgene_census.open_soma() as census:\n","    adata = cellxgene_census.get_anndata(\n","        census = census,\n","        organism = \"Homo sapiens\",\n","        var_value_filter = \"feature_id in ['ENSG00000161798', 'ENSG00000188229']\",\n","        obs_value_filter = \"sex == 'female' and cell_type in ['microglial cell', 'neuron']\",\n","        column_names = {\"obs\": [\"assay\", \"cell_type\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]},\n","    )\n","\n","    print(adata)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"I1qECth7mQMe","outputId":"69d18838-da49-407e-c1b3-0eac82bceb04","executionInfo":{"status":"ok","timestamp":1729576516288,"user_tz":480,"elapsed":812323,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}}},"execution_count":1,"outputs":[{"metadata":{"tags":null},"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.10/dist-packages/tiledb/cloud/config.py:96: UserWarning: You must first login before you can run commands. Please run tiledb.cloud.login.\n","  warnings.warn(\n","The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n","INFO:cellxgene_census:The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n","<ipython-input-1-31ea77b51fb3>:4: FutureWarning: The argument `column_names` is deprecated and will be removed in a future release. Please use `obs_column_names` and `var_column_names` instead.\n","  adata = cellxgene_census.get_anndata(\n"]},{"output_type":"stream","name":"stdout","text":["AnnData object with n_obs × n_vars = 732781 × 2\n","    obs: 'assay', 'cell_type', 'tissue', 'tissue_general', 'suspension_type', 'disease', 'sex'\n","    var: 'soma_joinid', 'feature_id', 'feature_name', 'feature_length', 'nnz', 'n_measured_obs'\n"]}]},{"cell_type":"markdown","source":["**Memory-efficient queries**"],"metadata":{"id":"9Zg-IuiIr7Ko"}},{"cell_type":"code","source":["import cellxgene_census\n","import tiledbsoma\n","\n","with cellxgene_census.open_soma() as census:\n","\n","    human = census[\"census_data\"][\"homo_sapiens\"]\n","    query = human.axis_query(\n","       measurement_name = \"RNA\",\n","       obs_query = tiledbsoma.AxisQuery(\n","           value_filter = \"tissue == 'brain' and sex == 'male'\"\n","       )\n","    )\n","    iterator = query.X(\"raw\").tables()\n","\n","    # Get an iterative slice as pyarrow.Table\n","    raw_slice = next (iterator)\n","    query.close()"],"metadata":{"id":"Ew8_a_O4sQrA","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1729576767698,"user_tz":480,"elapsed":127095,"user":{"displayName":"Kshitij Kakade","userId":"17584620299229098488"}},"outputId":"a505174d-2ac4-4836-ceff-ab822d20e190"},"execution_count":4,"outputs":[{"output_type":"stream","name":"stderr","text":["The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n","INFO:cellxgene_census:The \"stable\" release is currently 2024-07-01. Specify 'census_version=\"2024-07-01\"' in future calls to open_soma() to ensure data consistency.\n"]}]}]}